

<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>RTG Command Reference &#8212; RTG Tools Operations Manual v3.7</title>
    
    <link rel="stylesheet" href="_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '3.7',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/bizstyle.js"></script>
    <link rel="top" title="RTG Tools Operations Manual v3.7" href="index.html" />
    <link rel="next" title="RTG product usage - baseline progressions" href="product_usage_wrapper.html" />
    <link rel="prev" title="Overview" href="overview.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="product_usage_wrapper.html" title="RTG product usage - baseline progressions"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="overview.html" title="Overview"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">RTG Tools Operations Manual v3.7</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/html_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">RTG Command Reference</a><ul>
<li><a class="reference internal" href="#command-line-interface-cli">Command line interface (CLI)</a></li>
<li><a class="reference internal" href="#rtg-command-syntax">RTG command syntax</a></li>
<li><a class="reference internal" href="#data-formatting-commands">Data Formatting Commands</a><ul>
<li><a class="reference internal" href="#format">format</a></li>
<li><a class="reference internal" href="#sdf2fasta">sdf2fasta</a></li>
<li><a class="reference internal" href="#sdf2fastq">sdf2fastq</a></li>
<li><a class="reference internal" href="#sdf2sam">sdf2sam</a></li>
</ul>
</li>
<li><a class="reference internal" href="#utility-commands">Utility Commands</a><ul>
<li><a class="reference internal" href="#bgzip">bgzip</a></li>
<li><a class="reference internal" href="#index">index</a></li>
<li><a class="reference internal" href="#extract">extract</a></li>
<li><a class="reference internal" href="#aview">aview</a></li>
<li><a class="reference internal" href="#sdfstats">sdfstats</a></li>
<li><a class="reference internal" href="#sdfsubset">sdfsubset</a></li>
<li><a class="reference internal" href="#sdfsubseq">sdfsubseq</a></li>
<li><a class="reference internal" href="#mendelian">mendelian</a></li>
<li><a class="reference internal" href="#vcfstats">vcfstats</a></li>
<li><a class="reference internal" href="#vcfmerge">vcfmerge</a></li>
<li><a class="reference internal" href="#vcffilter">vcffilter</a></li>
<li><a class="reference internal" href="#vcfannotate">vcfannotate</a></li>
<li><a class="reference internal" href="#vcfsubset">vcfsubset</a></li>
<li><a class="reference internal" href="#vcfeval">vcfeval</a></li>
<li><a class="reference internal" href="#pedfilter">pedfilter</a></li>
<li><a class="reference internal" href="#pedstats">pedstats</a></li>
<li><a class="reference internal" href="#rocplot">rocplot</a></li>
<li><a class="reference internal" href="#version">version</a></li>
<li><a class="reference internal" href="#license">license</a></li>
<li><a class="reference internal" href="#help">help</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="overview.html"
                        title="previous chapter">Overview</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="product_usage_wrapper.html"
                        title="next chapter">RTG product usage - baseline progressions</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/rtg_command_reference.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="rtg-command-reference">
<h1>RTG Command Reference<a class="headerlink" href="#rtg-command-reference" title="Permalink to this headline">¶</a></h1>
<p>This chapter describes RTG commands with a generic description of
parameter options and usage. This section also includes expected
operation and output results.</p>
<div class="section" id="command-line-interface-cli">
<h2>Command line interface (CLI)<a class="headerlink" href="#command-line-interface-cli" title="Permalink to this headline">¶</a></h2>
<p>RTG is installed as a single executable in any system subdirectory where
permissions authorize a particular community of users to run the
application. RTG commands are executed through the RTG command-line
interface (CLI). Each command has its own set of parameters and options
described in this section. The availability of each command may be
determined by the RTG license that has been installed. Contact
<a class="reference external" href="mailto:support&#37;&#52;&#48;realtimegenomics&#46;com">support<span>&#64;</span>realtimegenomics<span>&#46;</span>com</a> to discuss changing the set of commands
that are enabled by your license.</p>
<p>Results are organized in results directories defined by command
parameters and settings. The command line shell environment should
include a set of familiar text post-processing tools, such as <code class="docutils literal"><span class="pre">grep</span></code>,
<code class="docutils literal"><span class="pre">awk</span></code>, or <code class="docutils literal"><span class="pre">perl</span></code>. Otherwise, no additional applications such as
databases or directory services are required.</p>
</div>
<div class="section" id="rtg-command-syntax">
<h2>RTG command syntax<a class="headerlink" href="#rtg-command-syntax" title="Permalink to this headline">¶</a></h2>
<p><strong>Usage:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>rtg COMMAND [OPTIONS] &lt;REQUIRED&gt;
</pre></div>
</div>
<p>To run an RTG command at the command prompt (either DOS window or Unix
terminal), type the product name followed by the command and all
required and optional parameters. For example:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format -o human_REF_SDF human_REF.fasta
</pre></div>
</div>
<p>Typically results are written to output files specified with the <code class="docutils literal"><span class="pre">-o</span></code>
option. There is no default filename or filename extension added to
commands requiring specification of an output directory or format.</p>
<p>Many times, unfiltered output files are very large; the built-in
compression option generates block compressed output files with the
.<code class="docutils literal"><span class="pre">gz</span></code> extension automatically unless the parameter <code class="docutils literal"><span class="pre">-Z</span></code> or <code class="docutils literal"><span class="pre">--no-gzip</span></code>
is issued with the command.</p>
<p>Many command parameters require user-supplied information of various
types, as shown in the following:</p>
<table border="1" class="docutils">
<colgroup>
<col width="17%" />
<col width="83%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Type</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>DIR, FILE</td>
<td>File or directory name(s)</td>
</tr>
<tr class="row-odd"><td>SDF</td>
<td>Sequence data that has been formatted to SDF</td>
</tr>
<tr class="row-even"><td>INT</td>
<td>Integer value</td>
</tr>
<tr class="row-odd"><td>FLOAT</td>
<td>Floating point decimal value</td>
</tr>
<tr class="row-even"><td>STRING</td>
<td>A sequence of characters for comments, filenames, or labels</td>
</tr>
</tbody>
</table>
<p>To display all parameters and syntax associated with an RTG command,
enter the command and type <code class="docutils literal"><span class="pre">--help</span></code>. For example: all parameters
available for the RTG <code class="docutils literal"><span class="pre">format</span></code> command are displayed when <code class="docutils literal"><span class="pre">rtg</span> <span class="pre">format</span>
<span class="pre">--help</span></code> is executed, the output of which is shown below.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>Usage: rtg format [OPTION]... -o SDF FILE+
                  [OPTION]... -o SDF -I FILE
                  [OPTION]... -o SDF -l FILE -r FILE

Converts the contents of sequence data files (FASTA/FASTQ/SAM/BAM) into the RTG
Sequence Data File (SDF) format.

File Input/Output
  -f, --format=FORMAT            format of input. Allowed values are [fasta,
                                 fastq, sam-se, sam-pe, cg-fastq, cg-sam]
                                 (Default is fasta)
  -I, --input-list-file=FILE     file containing a list of input read files (1
                                 per line)
  -l, --left=FILE                left input file for FASTA/FASTQ paired end
                                 data
  -o, --output=SDF               name of output SDF
  -p, --protein                  input is protein. If this option is not
                                 specified, then the input is assumed to
                                 consist of nucleotides
  -q, --quality-format=FORMAT    format of quality data for fastq files (use
                                 sanger for Illumina 1.8+). Allowed values are
                                 [sanger, solexa, illumina]
  -r, --right=FILE               right input file for FASTA/FASTQ paired end
                                 data
      FILE+                      input sequence files. May be specified 0 or
                                 more times

Filtering
      --duster                   treat lower case residues as unknowns
      --exclude=STRING           exclude input sequences based on their name.
                                 If the input sequence contains the specified
                                 string then that sequence is excluded from the
                                 SDF. May be specified 0 or more times
      --select-read-group=STRING when formatting from SAM/BAM input, only
                                 include reads with this read group ID
      --trim-threshold=INT       trim read ends to maximise base quality above
                                 the given threshold

Utility
      --allow-duplicate-names    disable checking for duplicate sequence names
  -h, --help                     print help on command-line flag usage
      --no-names                 do not include name data in the SDF output
      --no-quality               do not include quality data in the SDF output
      --sam-rg=STRING|FILE       file containing a single valid read group SAM
                                 header line or a string in the form
                                 &quot;@RG\tID:READGROUP1\tSM:BACT_SAMPLE\tPL:ILLUMINA&quot;
</pre></div>
</div>
<p>Required parameters are indicated in the usage display; optional
parameters are listed immediately below the usage information in
organized categories.</p>
<p>Use the double-dash when typing the full-word command option, as in
<code class="docutils literal"><span class="pre">--output</span></code>:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format --output human_REF_SDF human_REF.fasta
</pre></div>
</div>
<p>Commonly used command options provide an abbreviated single-character
version of a full command parameter, indicated with only a single dash,
(Thus <code class="docutils literal"><span class="pre">--output</span></code> is the same as specifying the command option with the
abbreviated character <code class="docutils literal"><span class="pre">-o</span></code>):</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format -o human_REF human_REF.fasta
</pre></div>
</div>
<p>A set of utility commands are provided through the CLI: <code class="docutils literal"><span class="pre">version</span></code>,
<code class="docutils literal"><span class="pre">license</span></code>, and <code class="docutils literal"><span class="pre">help</span></code>. Start with these commands to familiarize yourself
with the software.</p>
<p>The <code class="docutils literal"><span class="pre">rtg</span> <span class="pre">version</span></code> command invokes the RTG software and triggers the
launch of RTG product commands, options, and utilities:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg version
</pre></div>
</div>
<p>It will display the version of the RTG software installed, RAM
requirements, and license expiration, for example:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$rtg version
Product: RTG Core 3.5
Core Version: 6236f4e (2014-10-31)
RAM: 40.0GB of 47.0GB RAM can be used by rtg (84%)
License: Expires on 2015-09-30
License location: /home/rtgcustomer/rtg/rtg-license.txt
Contact: support@realtimegenomics.com

Patents / Patents pending:
US: 7,640,256, 13/129,329, 13/681,046, 13/681,215, 13/848,653,
13/925,704, 14/015,295, 13/971,654, 13/971,630, 14/564,810
UK: 1222923.3, 1222921.7, 1304502.6, 1311209.9, 1314888.7, 1314908.3
New Zealand: 626777, 626783, 615491, 614897, 614560
Australia: 2005255348, Singapore: 128254

Citation:
John G. Cleary, Ross Braithwaite, Kurt Gaastra, Brian S. Hilbush, Stuart
Inglis, Sean A. Irvine, Alan Jackson, Richard Littin, Sahar
Nohzadeh-Malakshah, Mehul Rathod, David Ware, Len Trigg, and Francisco
M. De La Vega. &quot;Joint Variant and De Novo Mutation Identification on
Pedigrees from High-Throughput Sequencing Data.&quot; Journal of
Computational Biology. June 2014, 21(6): 405-419.
doi:10.1089/cmb.2014.0029.
(c) Real Time Genomics Inc, 2014
</pre></div>
</div>
<p>To see what commands you are licensed to use, type <code class="docutils literal"><span class="pre">rtg</span> <span class="pre">license</span></code>:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$rtg license
License: Expires on 2015-03-30
Licensed to: John Doe
License location: /home/rtgcustomer/rtg/rtg-license.txt

    Command name    Licensed?  Release Level

Data formatting:
    format          Licensed   GA
    sdf2fasta       Licensed   GA
    sdf2fastq       Licensed   GA

Utility:
    bgzip           Licensed   GA
    index           Licensed   GA
    extract         Licensed   GA
    sdfstats        Licensed   GA
    sdfsubset       Licensed   GA
    sdfsubseq       Licensed   GA
    mendelian       Licensed   GA
    vcfstats        Licensed   GA
    vcfmerge        Licensed   GA
    vcffilter       Licensed   GA
    vcfannotate     Licensed   GA
    vcfsubset       Licensed   GA
    vcfeval         Licensed   GA
    pedfilter       Licensed   GA
    pedstats        Licensed   GA
    rocplot         Licensed   GA
    version         Licensed   GA
    license         Licensed   GA
    help            Licensed   GA
</pre></div>
</div>
<p>To display all commands and usage parameters available to use with your
license, type <code class="docutils literal"><span class="pre">rtg</span> <span class="pre">help</span></code>:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg help
Usage: rtg COMMAND [OPTION]...
       rtg RTG_MEM=16G COMMAND [OPTION]...  (e.g. to set maximum memory use to 16 GB)

Type &#39;rtg help COMMAND&#39; for help on a specific command. The
following commands are available:

Data formatting:
      format                 convert a FASTA file to SDF
      cg2sdf                 convert Complete Genomics reads to SDF
      sdf2fasta              convert SDF to FASTA
      sdf2fastq              convert SDF to FASTQ
      sdf2sam                convert SDF to SAM/BAM
Read mapping:
      map                    read mapping
      mapf                   read mapping for filtering purposes
      cgmap                  read mapping for Complete Genomics data
Protein search:
      mapx                   translated protein search
Assembly:
      assemble               assemble reads into long sequences
      addpacbio              add Pacific Biosciences reads to an assembly
Variant detection:
      calibrate              create calibration data from SAM/BAM files
      svprep                 prepare SAM/BAM files for sv analysis
      sv                     find structural variants
      discord                detect structural variant breakends using discordant reads
      coverage               calculate depth of coverage from SAM/BAM files
      snp                    call variants from SAM/BAM files
      family                 call variants for a family following Mendelian inheritance
      somatic                call variants for a tumor/normal pair
      population             call variants for multiple potentially-related individuals
      lineage                call de novo variants in a cell lineage
      avrbuild               AVR model builder
      avrpredict             run AVR on a VCF file
      cnv                    call CNVs from paired SAM/BAM files
Metagenomics:
      species                estimate species frequency in metagenomic samples
      similarity             calculate similarity matrix and nearest neighbor tree
Simulation:
      genomesim              generate simulated genome sequence
      cgsim                  generate simulated reads from a sequence
      readsim                generate simulated reads from a sequence
      readsimeval            evaluate accuracy of mapping simulated reads
      popsim                 generate a VCF containing simulated population variants
      samplesim              generate a VCF containing a genotype simulated from a population
      childsim               generate a VCF containing a genotype simulated as a child of two parents
      denovosim              generate a VCF containing a derived genotype containing de novo variants
      samplereplay           generate the genome corresponding to a sample genotype
      cnvsim                 generate a mutated genome by adding CNVs to a template
Utility:
      bgzip                  compress a file using block gzip
      index                  create a tabix index
      extract                extract data from a tabix indexed file
      sdfstats               print statistics about an SDF
      sdfsplit               split an SDF into multiple parts
      sdfsubset              extract a subset of an SDF into a new SDF
      sdfsubseq              extract a subsequence from an SDF as text
      sam2bam                convert SAM file to BAM file and create index
      sammerge               merge sorted SAM/BAM files
      samstats               print statistics about a SAM/BAM file
      samrename              rename read id to read name in SAM/BAM files
      mapxrename             rename read id to read name in mapx output files
      mendelian              check a multi-sample VCF for Mendelian consistency
      vcfstats               print statistics from about variants contained within a VCF file
      vcfmerge               merge single-sample VCF files into a single multi-sample VCF
      vcffilter              filter records within a VCF file
      vcfannotate            annotate variants within a VCF file
      vcfsubset              create a VCF file containing a subset of the original columns
      vcfeval                evaluate called variants for agreement with a baseline variant set
      pedfilter              filter and convert a pedigree file
      pedstats               print information about a pedigree file
      avrstats               print statistics about an AVR model
      rocplot                plot ROC curves from vcfeval ROC data files
      usageserver            run a local server for collecting RTG command usage information
      version                print version and license information
      license                print license information for all commands
      help                   print this screen or help for specified command
</pre></div>
</div>
<p>The help command will only list the commands for which you have a
license to use.</p>
<p>To display help and syntax information for a specific command from the
command line, type the command and then the &#8211;help option, as in:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format --help
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The following commands are synonymous:
<code class="docutils literal"><span class="pre">rtg</span> <span class="pre">help</span> <span class="pre">format</span></code> and <code class="docutils literal"><span class="pre">rtg</span> <span class="pre">format</span> <span class="pre">--help</span></code></p>
</div>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">Refer to <a class="reference internal" href="overview.html#installation-and-deployment"><span class="std std-ref">Installation and deployment</span></a> for information
about installing the RTG product executable.</p>
</div>
</div>
<div class="section" id="data-formatting-commands">
<h2>Data Formatting Commands<a class="headerlink" href="#data-formatting-commands" title="Permalink to this headline">¶</a></h2>
<div class="section" id="format">
<h3>format<a class="headerlink" href="#format" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>The <code class="docutils literal"><span class="pre">format</span></code> command converts the contents of sequence data files
(FASTA/FASTQ/SAM/BAM) into the RTG Sequence Data File (SDF) format. This
step ensures efficient processing of very large data sets, by organizing
the data into multiple binary files within a named directory. The same
SDF format is used for storing sequence data, whether it be genomic
reference, sequencing reads, protein sequences, etc.</p>
<p><strong>Syntax:</strong></p>
<p>Format one or more files specified from command line into a single SDF:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format [OPTION] -o SDF FILE+
</pre></div>
</div>
<p>Format one or more files specified in a text file into a single SDF:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format [OPTION] -o SDF -I FILE
</pre></div>
</div>
<p>Format mate pair reads into a single SDF:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format [OPTION] -o SDF -l FILE -r FILE
</pre></div>
</div>
<p><strong>Examples:</strong></p>
<p>For FASTA (.fa) genome reference data:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format -o maize_reference maize_chr*.fa
</pre></div>
</div>
<p>For FASTQ (.fq) sequence read data:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format -f fastq -q sanger -o h1_reads -l h1_sample_left.fq -r h1_sample_right.fq
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="18%" />
<col width="77%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-f</span></code></td>
<td><code class="docutils literal"><span class="pre">--format=FORMAT</span></code></td>
<td>The format of the input file(s). Allowed values are [fasta, fastq, sam-se, sam-pe, cg-fastq, cg-sam] (Default is fasta).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-I</span></code></td>
<td><code class="docutils literal"><span class="pre">--input-list-file=FILE</span></code></td>
<td>Specifies a file containing a list of sequence data files (one per line) to be converted into an SDF.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-l</span></code></td>
<td><code class="docutils literal"><span class="pre">--left=FILE</span></code></td>
<td>The left input file for FASTA/FASTQ paired end data.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-o</span></code></td>
<td><code class="docutils literal"><span class="pre">--output=SDF</span></code></td>
<td>The name of the output SDF.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-p</span></code></td>
<td><code class="docutils literal"><span class="pre">--protein</span></code></td>
<td>Set if the input consists of protein. If this option is not specified, then the input is assumed to consist of nucleotides.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-q</span></code></td>
<td><code class="docutils literal"><span class="pre">--quality-format=FORMAT</span></code></td>
<td>The format of the quality data for fastq format files. (Use sanger for Illumina1.8+). Allowed values are [sanger, solexa, illumina].</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-r</span></code></td>
<td><code class="docutils literal"><span class="pre">--right=FILE</span></code></td>
<td>The right input file for FASTA/FASTQ paired end data.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">FILE+</span></code></td>
<td>Specifies a sequence data file to be converted into an SDF. May be specified 0 or more times.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="2%" />
<col width="18%" />
<col width="80%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Filtering</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--duster</span></code></td>
<td>Treat lower case residues as unknowns.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--exclude=STRING</span></code></td>
<td>Exclude individual input sequences based on their name. If the input sequence name contains the specified string then that sequence is excluded from the
SDF. May be specified 0 or more times.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--select-read-group=STRING</span></code></td>
<td>Set to only include only reads with this read group ID when formatting from SAM/BAM files.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--trim-threshold=INT</span></code></td>
<td>Set to trim the read ends to maximise the base quality above the given threshold.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="16%" />
<col width="79%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--allow-duplicate-names</span></code></td>
<td>Set to disable duplicate name detection.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--no-names</span></code></td>
<td>Do not include sequence names in the resulting SDF.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--no-quality</span></code></td>
<td>Do not include sequence quality data in the resulting SDF.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--sam-rg=STRING|FILE</span></code></td>
<td>Specifies a file containing a single valid read group SAM header line or a string in the form <code class="docutils literal"><span class="pre">&#64;RG\tID:RG1\tSM:G1_SAMP\tPL:ILLUMINA</span></code>.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Formatting takes one or more input data files and creates a single SDF.
Specify the type of file to be converted, or allow default to FASTA
format. To aggregate multiple input data files, such as when formatting
a reference genome consisting of multiple chromosomes, list all files on
the command line or use the <code class="docutils literal"><span class="pre">--input-list-file</span></code> flag to specify a file
containing the list of files to process.</p>
<p>For input FASTA and FASTQ files which are compressed, they must have a
filename extension of <code class="docutils literal"><span class="pre">.gz</span></code> (for gzip compressed data) or <code class="docutils literal"><span class="pre">.bz2</span></code> (for
bzip2 compressed data).</p>
<p>When formatting human reference genome data, it is recommended that the
resulting SDF be augmented with chromosome reference metadata, in order
to enable automatic sex-aware features during mapping and variant
calling. The <code class="docutils literal"><span class="pre">format</span></code> command will automatically recognize several
common human reference genomes and install a reference configuration
file. If your reference genome is not recognized, a configuration can be
manually adapted from one of the examples provided in the RTG
distribution and installed in the SDF directory. The reference
configuration is described in <a class="reference internal" href="appendix.html#rtg-reference-file-format"><span class="std std-ref">RTG reference file format</span></a>.</p>
<p>When using FASTQ input files you must specify the quality format being
used as one of <code class="docutils literal"><span class="pre">sanger</span></code>, <code class="docutils literal"><span class="pre">solexa</span></code> or <code class="docutils literal"><span class="pre">illumina</span></code>. As of Illumina pipeline
version 1.8 and higher, quality values are encoded in Sanger format and
so should be formatted using <code class="docutils literal"><span class="pre">--quality-format=sanger</span></code>. Output from
earlier Illumina pipeline versions should be formatted using
<code class="docutils literal"><span class="pre">--quality-format=illumina</span></code> for Illumina pipeline versions starting with
1.3 and before 1.8, or <code class="docutils literal"><span class="pre">--quality-format=solexa</span></code> for Illumina pipeline
versions less than 1.3.</p>
<p>For files that represent paired-end read data, indicate each side
respectively using the <code class="docutils literal"><span class="pre">--left=FILE</span></code> and <code class="docutils literal"><span class="pre">--right=FILE</span></code> flags.</p>
<p>The <code class="docutils literal"><span class="pre">mapx</span></code> command maps translated DNA sequence data against a protein
reference. You must use the <code class="docutils literal"><span class="pre">-p</span></code>, <code class="docutils literal"><span class="pre">--protein</span></code> flag to format the protein
reference used by <code class="docutils literal"><span class="pre">mapx</span></code>.</p>
<p>Use the <code class="docutils literal"><span class="pre">sam-se</span></code> format for single end SAM/BAM input files and the
<code class="docutils literal"><span class="pre">sam-pe</span></code> format for paired end SAM/BAM input files. Note that if the
input SAM/BAM files are sorted in coordinate order (for example if they
have already been aligned to a reference), it is recommended that they
be shuffled before formatting, so that subsequent mapping is not biased
by processing reads in chromosome order. For example, a BAM file can be
shuffled using <code class="docutils literal"><span class="pre">samtools</span> <span class="pre">bamshuf</span></code> as follows:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ samtools bamshuf -uOn 256 reads.bam tmp-prefix &gt;reads_shuffled.bam
</pre></div>
</div>
<p>And this can be carried out on the fly during formatting using bash
process redirection in order to reduce intermediate I/O, for example:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format --format sam-pe &lt;(samtools bamshuf -uOn 256 reads.bam temp-prefix) ...
</pre></div>
</div>
<p>The SDF for a read set can contain a SAM read group which will be
automatically picked up from the input SAM/BAM files if they contain
only one read group. If the input SAM/BAM files contain multiple read
groups you must select a single read group from the SAM/BAM file to
format using the <code class="docutils literal"><span class="pre">--select-read-group</span></code> flag or specify a custom read
group with the <code class="docutils literal"><span class="pre">--sam-rg</span></code> flag. The <code class="docutils literal"><span class="pre">--sam-rg</span></code> flag can also be used to
add read group information to reads given in other input formats. The
SAM read group stored in an SDF will be automatically used during
mapping the reads it contains to provide tracking information in the
output BAM files.</p>
<p>The <code class="docutils literal"><span class="pre">--trim-threshold</span></code> flag can be used to trim poor quality read ends
from the input reads by inspecting base qualities from FASTQ input. If
and only if the quality of the final base of the read is less than the
threshold given, a new read length is found which maximizes the overall
quality of the retained bases using the following formula.</p>
<div class="math">
<p><img src="_images/math/d87c941fdcf46735e757ab196fe03da2ea08ec45.png" alt="\arg \max x\left({\sum_{i=x+1}^l (T - q(i))}\right) \text{ if } q(l) &lt; T"/></p>
</div><p>Where <em>l</em> is the original read length, <em>x</em> is the new read length, <em>T</em>
is the given threshold quality and <em>q(n)</em> is the quality of the base at
the position n of the read.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>Sequencing system read files and reference genome files
often have the same extension and it may not always be obvious which
file is a read set and which is a genome. Before formatting a sequencing
system file, open it to see which type of file it is. For example:</p>
<div class="last highlight-text"><div class="highlight"><pre><span></span>$ less pf3.fa
</pre></div>
</div>
</div>
<p>In general, a read file typically begins with an <code class="docutils literal"><span class="pre">&#64;</span></code> or <code class="docutils literal"><span class="pre">+</span></code> character; a
genome reference file typically begins with the characters <code class="docutils literal"><span class="pre">chr</span></code>.</p>
<p>Normally when the input data contains multiple sequences with the same name the
format command will fail with an error. The <code class="docutils literal"><span class="pre">--allow-duplicate-names</span></code> flag
will disable this check conserving memory, but if the input data has multiple
sequences with the same name you will not be warned. Having duplicate sequence
names can cause problems with other commands, especially for reference data
since the output many commands identifies sequences by their names.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#sdf2fasta"><span class="std std-ref">sdf2fasta</span></a>,
<a class="reference internal" href="#sdf2fastq"><span class="std std-ref">sdf2fastq</span></a>,
<a class="reference internal" href="#sdfstats"><span class="std std-ref">sdfstats</span></a>,
<a class="reference internal" href="#cg2sdf"><span class="std std-ref">cg2sdf</span></a>,
<a class="reference internal" href="#map"><span class="std std-ref">map</span></a>,
<a class="reference internal" href="#sdfsplit"><span class="std std-ref">sdfsplit</span></a></p>
</div>
</div>
<div class="section" id="sdf2fasta">
<h3>sdf2fasta<a class="headerlink" href="#sdf2fasta" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Convert SDF data into a FASTA file.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdf2fasta [OPTION]... -i SDF -o FILE
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdf2fasta -i humanSDF -o humanFASTA_return
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="10%" />
<col width="21%" />
<col width="69%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-i</span></code></td>
<td><code class="docutils literal"><span class="pre">--input=SDF</span></code></td>
<td>Specifies the SDF data to be converted.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-o</span></code></td>
<td><code class="docutils literal"><span class="pre">--output=FILE</span></code></td>
<td>Specifies the file name used to write the resulting FASTA output.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="12%" />
<col width="82%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Filtering</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--end-id=INT</span></code></td>
<td>Only output sequences with sequence id less than the given number. (Sequence ids start at 0).</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--start-id=INT</span></code></td>
<td>Only output sequences with sequence id greater than or equal to the given number. (Sequence ids start at 0).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-I</span></code></td>
<td><code class="docutils literal"><span class="pre">--id-file=FILE</span></code></td>
<td>Name of a file containing a list of sequences to extract, one per line.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--names</span></code></td>
<td>Interpret any specified sequence as names instead of numeric sequence ids.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--taxons</span></code></td>
<td>Interpret any specified sequence as taxon ids instead of numeric sequence ids. This option only applies to a metagenomic reference species SDF.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">STRING+</span></code></td>
<td>Specify one or more explicit sequences to extract, as sequence id, or sequence name if &#8211;names flag is set.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="13%" />
<col width="82%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--interleave</span></code></td>
<td>Interleave paired data into a single output file. Default is to split to separate output files.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-l</span></code></td>
<td><code class="docutils literal"><span class="pre">--line-length=INT</span></code></td>
<td>Set the maximum number of nucleotides or amino acids to print on a line of FASTA output. Should be nonnegative, with a value of 0 indicating that the line
length is not capped. (Default is 0).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-Z</span></code></td>
<td><code class="docutils literal"><span class="pre">--no-gzip</span></code></td>
<td>Set this flag to create the FASTA output file without compression. By default the output file is compressed with blocked gzip.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Use the <code class="docutils literal"><span class="pre">sdf2fasta</span></code> command to convert SDF data into FASTA format. By
default, <code class="docutils literal"><span class="pre">sdf2fasta</span></code> creates a separate line of FASTA output for each
sequence. These lines will be as long as the sequences themselves. To
make them more readable, use the <code class="docutils literal"><span class="pre">-l</span></code>, <code class="docutils literal"><span class="pre">--line-length</span></code> flag and define a
reasonable record length like 75.</p>
<p>By default all sequences will be extracted, but flags may be specified
to extract reads within a range, or explicitly specified reads (either
by numeric sequence id or by sequence name if <code class="docutils literal"><span class="pre">--names</span></code> is set).
Additionally, when the input SDF is a metagenomic species reference SDF,
the <code class="docutils literal"><span class="pre">--taxons</span></code> option, any supplied id is interpreted as a taxon id and
all sequences assigned directly to that taxon id will be output. This
provides a convenient way to extract all sequence data corresponding to
a single (or multiple) species from a metagenomic species reference SDF.</p>
<p>Sequence ids are numbered starting at 0, the <code class="docutils literal"><span class="pre">--start-id</span></code> flag is an inclusive
lower bound on id and the <code class="docutils literal"><span class="pre">--end-id</span></code> flag is an exclusive upper bound. For
example if you have an SDF with five sequences (ids: 0, 1, 2, 3, 4) the
following command:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdf2fasta --start-id=3 -i mySDF -o output
</pre></div>
</div>
<p>will extract sequences with id 3 and 4. The command:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdf2fasta --end-id=3 -i mySDF -o output
</pre></div>
</div>
<p>will extract sequences with id 0, 1, and 2. And the command:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdf2fasta --start-id=2 --end-id=4 -i mySDF -o output
</pre></div>
</div>
<p>will extract sequences with id 2 and 3.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#format"><span class="std std-ref">format</span></a>,
<a class="reference internal" href="#sdf2fastq"><span class="std std-ref">sdf2fastq</span></a>,
<a class="reference internal" href="#sdfstats"><span class="std std-ref">sdfstats</span></a>,
<a class="reference internal" href="#cg2sdf"><span class="std std-ref">cg2sdf</span></a>,
<a class="reference internal" href="#sdfsplit"><span class="std std-ref">sdfsplit</span></a></p>
</div>
</div>
<div class="section" id="sdf2fastq">
<h3>sdf2fastq<a class="headerlink" href="#sdf2fastq" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Convert SDF data into a FASTQ file.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdf2fastq [OPTION]... -i SDF -o FILE
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdf2fastq -i humanSDF -o humanFASTQ_return
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="10%" />
<col width="21%" />
<col width="69%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-i</span></code></td>
<td><code class="docutils literal"><span class="pre">--input=SDF</span></code></td>
<td>Specifies the SDF data to be converted.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-o</span></code></td>
<td><code class="docutils literal"><span class="pre">--output=FILE</span></code></td>
<td>Specifies the file name used to write the resulting FASTQ output.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="15%" />
<col width="78%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Filtering</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--end-id=INT</span></code></td>
<td>Only output sequences with sequence id less than the given number. (Sequence ids start at 0).</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--start-id=INT</span></code></td>
<td>Only output sequences with sequence id greater than or equal to the given number. (Sequence ids start at 0).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-I</span></code></td>
<td><code class="docutils literal"><span class="pre">--id-file=FILE</span></code></td>
<td>Name of a file containing a list of sequences to extract, one per line.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--names</span></code></td>
<td>Interpret any specified sequence as names instead of numeric sequence ids.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">STRING+</span></code></td>
<td>Specify one or more explicit sequences to extract, as sequence id, or sequence name if &#8211;names flag is set.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="14%" />
<col width="80%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-q</span></code></td>
<td><code class="docutils literal"><span class="pre">--default-qualty=INT</span></code></td>
<td>Set the default quality to use if the SDF does not contain sequence quality data (0-63).</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--interleave</span></code></td>
<td>Interleave paired data into a single output file. Default is to split to separate output files.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-l</span></code></td>
<td><code class="docutils literal"><span class="pre">--line-length=INT</span></code></td>
<td>Set the maximum number of nucleotides or amino acids to print on a line of FASTQ output. Should be nonnegative, with a value of 0 indicating that the line
length is not capped. (Default is 0).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-Z</span></code></td>
<td><code class="docutils literal"><span class="pre">--no-gzip</span></code></td>
<td>Set this flag to create the FASTQ output file without compression. By default the output file is compressed with blocked gzip.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Use the <code class="docutils literal"><span class="pre">sdf2fastq</span></code> command to convert SDF data into FASTQ format. If no
quality data is available in the SDF, use the <code class="docutils literal"><span class="pre">-q</span></code>, <code class="docutils literal"><span class="pre">--default-quality</span></code>
flag to set a quality score for the FASTQ output. The quality encoding
used during output is sanger quality encoding. By default, <code class="docutils literal"><span class="pre">sdf2fastq</span></code>
creates a separate line of FASTQ output for each sequence. As with
<code class="docutils literal"><span class="pre">sdf2fasta</span></code>, there is an option to use the <code class="docutils literal"><span class="pre">-l</span></code>, <code class="docutils literal"><span class="pre">--line-length</span></code> flag to
restrict the line lengths to improve readability of long sequences.</p>
<p>By default all sequences will be extracted, but flags may be specified
to extract reads within a range, or explicitly specified reads (either
by numeric sequence id or by sequence name if <code class="docutils literal"><span class="pre">--names</span></code> is set).</p>
<p>It may be preferable to extract data to unaligned SAM/BAM format using
<code class="docutils literal"><span class="pre">sdf2sam</span></code>, as this preserves read-group information stored in the SDF
and may also be more convenient when dealing with paired-end data.</p>
<p>The <code class="docutils literal"><span class="pre">--start-id</span></code> and <code class="docutils literal"><span class="pre">--end-id</span></code> flags behave as in <code class="docutils literal"><span class="pre">sdf2fasta</span></code>.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#format"><span class="std std-ref">format</span></a>,
<a class="reference internal" href="#sdf2fasta"><span class="std std-ref">sdf2fasta</span></a>,
<a class="reference internal" href="#sdf2sam"><span class="std std-ref">sdf2sam</span></a>,
<a class="reference internal" href="#sdfstats"><span class="std std-ref">sdfstats</span></a>,
<a class="reference internal" href="#cg2sdf"><span class="std std-ref">cg2sdf</span></a>,
<a class="reference internal" href="#sdfsplit"><span class="std std-ref">sdfsplit</span></a></p>
</div>
</div>
<div class="section" id="sdf2sam">
<h3>sdf2sam<a class="headerlink" href="#sdf2sam" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Convert SDF read data into unaligned SAM or BAM format file.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdf2sam [OPTION]... -i SDF -o FILE
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdf2sam -i samplereadsSDF -o samplereads.bam
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="13%" />
<col width="80%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-i</span></code></td>
<td><code class="docutils literal"><span class="pre">--input=SDF</span></code></td>
<td>Specifies the SDF data to be converted.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-o</span></code></td>
<td><code class="docutils literal"><span class="pre">--output=FILE</span></code></td>
<td>Specifies the file name used to write the resulting SAM/BAM to. The output format is automatically determined based on the
filename specified. If &#8216;-&#8216; is given, the data is written as uncompressed SAM to standard output.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="15%" />
<col width="78%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Filtering</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--end-id=INT</span></code></td>
<td>Only output sequences with sequence id less than the given number. (Sequence ids start at 0).</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--start-id=INT</span></code></td>
<td>Only output sequences with sequence id greater than or equal to the given number. (Sequence ids start at 0).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-I</span></code></td>
<td><code class="docutils literal"><span class="pre">--id-file=FILE</span></code></td>
<td>Name of a file containing a list of sequences to extract, one per line.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--names</span></code></td>
<td>Interpret any specified sequence as names instead of numeric sequence ids.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">STRING+</span></code></td>
<td>Specify one or more explicit sequences to extract, as sequence id, or sequence name if &#8211;names flag is set.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="10%" />
<col width="85%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-Z</span></code></td>
<td><code class="docutils literal"><span class="pre">--no-gzip</span></code></td>
<td>Set this flag when creating SAM format output to disable compression. By default SAM is compressed with blocked gzip, and BAM is always compressed.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Use the <code class="docutils literal"><span class="pre">sdf2sam</span></code> command to convert SDF data into unaligned SAM/BAM
format. By default all sequences will be extracted, but flags may be
specified to extract reads within a range, or explicitly specified reads
(either by numeric sequence id or by sequence name if <code class="docutils literal"><span class="pre">--names</span></code> is set).
This command is a useful way to export paired-end data to a single
output file while retaining any read group information that may be
stored in the SDF.</p>
<p>The output format is either SAM/BAM depending on the specified output file name.
e.g. <code class="docutils literal"><span class="pre">output.sam</span></code> or <code class="docutils literal"><span class="pre">output.sam.gz</span></code> will output as SAM, whereas
<code class="docutils literal"><span class="pre">output.bam</span></code> will output as BAM. If neither SAM or BAM format is indicated by
the file name then BAM will be used and the output file name adjusted
accordingly. e.g <code class="docutils literal"><span class="pre">output</span></code> will become <code class="docutils literal"><span class="pre">output.bam</span></code>. However if stardard
output is selected (<code class="docutils literal"><span class="pre">-</span></code>) then the output will always be in uncompressed SAM
format.</p>
<p>The <code class="docutils literal"><span class="pre">--start-id</span></code> and <code class="docutils literal"><span class="pre">--end-if</span></code> behave as in <code class="docutils literal"><span class="pre">sdf2fasta</span></code>.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#format"><span class="std std-ref">format</span></a>,
<a class="reference internal" href="#sdf2fasta"><span class="std std-ref">sdf2fasta</span></a>,
<a class="reference internal" href="#sdf2fastq"><span class="std std-ref">sdf2fastq</span></a>,
<a class="reference internal" href="#sdfstats"><span class="std std-ref">sdfstats</span></a>,
<a class="reference internal" href="#cg2sdf"><span class="std std-ref">cg2sdf</span></a>,
<a class="reference internal" href="#sdfsplit"><span class="std std-ref">sdfsplit</span></a></p>
</div>
</div>
</div>
<div class="section" id="utility-commands">
<h2>Utility Commands<a class="headerlink" href="#utility-commands" title="Permalink to this headline">¶</a></h2>
<div class="section" id="bgzip">
<h3>bgzip<a class="headerlink" href="#bgzip" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Block compress a file or decompress a block compressed file. Block
compressed outputs from the mapping and variant detection commands can
be indexed with the <code class="docutils literal"><span class="pre">index</span></code> command. They can also be processed with
standard gzip tools such as <code class="docutils literal"><span class="pre">gunzip</span></code> and <code class="docutils literal"><span class="pre">zcat</span></code>.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg bgzip [OPTION]... FILE+
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg bgzip alignments.sam
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="21%" />
<col width="72%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-l</span></code></td>
<td><code class="docutils literal"><span class="pre">--compression-level=INT</span></code></td>
<td>the compression level to use, between 1 (least but fast) and 9 (highest but slow) (Default is 5)</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-d</span></code></td>
<td><code class="docutils literal"><span class="pre">--decompress</span></code></td>
<td>Set to decompress the input file.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-f</span></code></td>
<td><code class="docutils literal"><span class="pre">--force</span></code></td>
<td>Overwrite the output file if it already exists.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--no-terminate</span></code></td>
<td>if set, do not add the block gzip termination block</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-c</span></code></td>
<td><code class="docutils literal"><span class="pre">--stdout</span></code></td>
<td>Write output to standard output, keep the original files unchanged. Implied when using standard input.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">FILE+</span></code></td>
<td>Specifies the file to be compressed or decompressed. Use
&#8216;-&#8216; to read from standard input. Must be specified 1 or more times.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="21%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Use the <code class="docutils literal"><span class="pre">bgzip</span></code> command to block compress files. Files such as VCF, BED,
SAM, TSV must be block-compressed before they can be indexed for fast
retrieval of records corresponding to specific genomic regions.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#index"><span class="std std-ref">index</span></a></p>
</div>
</div>
<div class="section" id="index">
<h3>index<a class="headerlink" href="#index" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Create tabix index files for block compressed TAB-delimited genome
position data files or BAM index files for BAM files.</p>
<p><strong>Syntax:</strong></p>
<p>Multi-file input specified from command line:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg index [OPTION]... -f FORMAT FILE+
</pre></div>
</div>
<p>Multi-file input specified in a text file:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg index [OPTION]... -f FORMAT -I FILE
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg index -f sam alignments.sam.gz
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="16%" />
<col width="78%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-f</span></code></td>
<td><code class="docutils literal"><span class="pre">--format=FORMAT</span></code></td>
<td>Specifies format of the input files to be indexed. Allowed values are [sam, bam, sv, coveragetsv, bed, vcf]</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-I</span></code></td>
<td><code class="docutils literal"><span class="pre">--input-list-file=FILE</span></code></td>
<td>Specifies a file containing a list of block compressed files (1 per line) containing data in the specified genome position format.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">FILE+</span></code></td>
<td>Specifies a block compressed file containing data in the specified genome position format to be indexed. May be specified 0 or more times.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="21%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Use the <code class="docutils literal"><span class="pre">index</span></code> command to produce tabix indexes for block compressed
genome position data files like SAM files and the output from <code class="docutils literal"><span class="pre">sv</span></code>,
<code class="docutils literal"><span class="pre">discord</span></code>, <code class="docutils literal"><span class="pre">coverage</span></code> and <code class="docutils literal"><span class="pre">snp</span></code> commands. The <code class="docutils literal"><span class="pre">index</span></code> command can also
be used to produce BAM indexes for BAM files with no index.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#map"><span class="std std-ref">map</span></a>,
<a class="reference internal" href="#coverage"><span class="std std-ref">coverage</span></a>,
<a class="reference internal" href="#snp"><span class="std std-ref">snp</span></a>,
<a class="reference internal" href="#sv"><span class="std std-ref">sv</span></a>,
<a class="reference internal" href="#discord"><span class="std std-ref">discord</span></a>,
<a class="reference internal" href="#extract"><span class="std std-ref">extract</span></a>,
<a class="reference internal" href="#bgzip"><span class="std std-ref">bgzip</span></a></p>
</div>
</div>
<div class="section" id="extract">
<h3>extract<a class="headerlink" href="#extract" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Extract specified parts of an indexed block compressed genome position
data file.</p>
<p><strong>Syntax:</strong></p>
<p>Extract whole file:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg extract [OPTION]... FILE
</pre></div>
</div>
<p>Extract specific regions:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg extract [OPTION]... FILE STRING+
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg extract alignments.bam &#39;chr1:10000+10&#39;
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="14%" />
<col width="81%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">FILE</span></code></td>
<td>The indexed block compressed genome position data file to extract.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="2%" />
<col width="8%" />
<col width="90%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Filtering</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">STRING+</span></code></td>
<td>Specifies the region to display. The format is one of &lt;sequence_name&gt;, &lt;sequence_name&gt;:start-end or &lt;sequence_name&gt;:start+length. May be specified 0 or more times.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="32%" />
<col width="62%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Reporting</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--header</span></code></td>
<td>Set to also display the file header.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--header-only</span></code></td>
<td>Set to only display the file header.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="21%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Use the <code class="docutils literal"><span class="pre">extract</span></code> command to view specific parts of indexed block
compressed genome position data files.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#map"><span class="std std-ref">map</span></a>,
<a class="reference internal" href="#coverage"><span class="std std-ref">coverage</span></a>,
<a class="reference internal" href="#snp"><span class="std std-ref">snp</span></a>,
<a class="reference internal" href="#sv"><span class="std std-ref">sv</span></a>,
<a class="reference internal" href="#index"><span class="std std-ref">index</span></a>,
<a class="reference internal" href="#bgzip"><span class="std std-ref">bgzip</span></a></p>
</div>
</div>
<div class="section" id="aview">
<h3>aview<a class="headerlink" href="#aview" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>View read mapping and variants corresponding to a region of the genome,
with output as ASCII to the terminal, or HTML.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg aview [OPTION]... --region STRING -t SDF FILE+
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg aview -t hg19 -b omni.vcf -c calls.vcf map/alignments.bam \
  --region Chr10:100000+3 –padding 30
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="20%" />
<col width="74%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-b</span></code></td>
<td><code class="docutils literal"><span class="pre">--baseline=FILE</span></code></td>
<td>VCF file containing baseline variants</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-B</span></code></td>
<td><code class="docutils literal"><span class="pre">--bed=FILE</span></code></td>
<td>BED file containing regions to overlay. May be specified 0 or more times.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-c</span></code></td>
<td><code class="docutils literal"><span class="pre">--calls=FILE</span></code></td>
<td>VCF file containing called variants. May be specified 0 or more times.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-I</span></code></td>
<td><code class="docutils literal"><span class="pre">--input-list-file=FILE</span></code></td>
<td>file containing a list of SAM/BAM format files (1 per line).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-r</span></code></td>
<td><code class="docutils literal"><span class="pre">--reads=SDF</span></code></td>
<td>read SDF (only needed to indicate correctness of simulated read mappings). May be specified 0 or more times.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-t</span></code></td>
<td><code class="docutils literal"><span class="pre">--template=SDF</span></code></td>
<td>reference SDF to which mappings and variants apply</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">FILE+</span></code></td>
<td>The indexed block compressed genome position data file to extract.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="11%" />
<col width="84%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Filtering</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-p</span></code></td>
<td><code class="docutils literal"><span class="pre">--padding=INT</span></code></td>
<td>Padding around region of interest. Default is to automatically determine padding to avoid read truncation.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--region=STRING</span></code></td>
<td>Specifies the region to display. The format is one of &lt;sequence_name&gt;, &lt;sequence_name&gt;:start-end or &lt;sequence_name&gt;:start+length. May be specified 0 or more times.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--sample=STRING</span></code></td>
<td>Specify name of sample to select. May be specified 0 or more times, or as a comma separated list</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="3%" />
<col width="22%" />
<col width="75%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Reporting</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--html</span></code></td>
<td>Output as HTML.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--no-base-colors</span></code></td>
<td>Do not use base-colors</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--no-color</span></code></td>
<td>Do not use colors.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--no-dots</span></code></td>
<td>Display nucleotide instead of dots.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--print-cigars</span></code></td>
<td>Print alignment cigars.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--print-mapq</span></code></td>
<td>Print alignment MAPQ values.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--print-mate-position</span></code></td>
<td>Print mate position.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--print-names</span></code></td>
<td>Print read names.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--print-readgroup</span></code></td>
<td>Print read group id for each alignment.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--print-reference-line=INT</span></code></td>
<td>print reference line every N lines (Default is 0).</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--print-soft-clipped-bases</span></code></td>
<td>Print soft clipped bases</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--project-track=INT</span></code></td>
<td>If set, project highlighting for the specified track down through reads. Default projects the union of tracks.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--sort-readgroup</span></code></td>
<td>Sort reads first on read group and then on start position.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--sort-reads</span></code></td>
<td>Sort reads on start position.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--unflatten</span></code></td>
<td>Display unflattened CGI reads when present</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="21%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Use the <code class="docutils literal"><span class="pre">aview</span></code> command to display a textual view of mappings and
variants corresponding to a small region of the reference genome. This
is useful when examining evidence for variant calls in a server
environment where a graphical display application such as IGV is not
available. The <code class="docutils literal"><span class="pre">aview</span></code> command is easy to script in order to output
displays for multiple regions for later viewing (either as text or
HTML).</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#map"><span class="std std-ref">map</span></a>,
<a class="reference internal" href="#snp"><span class="std std-ref">snp</span></a></p>
</div>
</div>
<div class="section" id="sdfstats">
<h3>sdfstats<a class="headerlink" href="#sdfstats" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Print statistics that describe a directory of SDF formatted data.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfstats [OPTION]... SDF+
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfstats human_READS_SDF

Location           : C:\human_READS_SDF
Parameters         : format -f solexa -o human_READS_SDF
                                c:\users\Elle\human\SRR005490.fastq.gz
SDF Version        : 6
Type               : DNA
Source             : SOLEXA
Paired arm         : UNKNOWN
Number of sequences: 4193903
Maximum length     : 48
Minimum length     : 48
N                  : 931268
A                  : 61100096
C                  : 41452181
G                  : 45262380
T                  : 52561419
Total residues     : 201307344
Quality scores available on this SDF
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="4%" />
<col width="11%" />
<col width="85%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">SDF+</span></code></td>
<td>Specifies an SDF on which statistics are to be reported. May be specified 1 or more times.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="9%" />
<col width="86%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Reporting</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--lengths</span></code></td>
<td>Set to print out the name and length of each sequence. (Not recommended for read sets).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-p</span></code></td>
<td><code class="docutils literal"><span class="pre">--position</span></code></td>
<td>Set to include information about unknown bases (Ns) by read position.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-q</span></code></td>
<td><code class="docutils literal"><span class="pre">--quality</span></code></td>
<td>Set to display mean of quality.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--sex=SEX</span></code></td>
<td>Set to display the reference sequence list for the given sex. Allowed values are [male, female, either]. May be specified 0 or more times, or as a comma separated list.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--taxonomy</span></code></td>
<td>Set to display information about the taxonomy.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-n</span></code></td>
<td><code class="docutils literal"><span class="pre">--unknowns</span></code></td>
<td>Set to include information about unknown bases (Ns).</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="21%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Use the <code class="docutils literal"><span class="pre">sdfstats</span></code> command to get information about the contents of
SDFs.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#format"><span class="std std-ref">format</span></a>,
<a class="reference internal" href="#sdf2fasta"><span class="std std-ref">sdf2fasta</span></a>,
<a class="reference internal" href="#sdf2fastq"><span class="std std-ref">sdf2fastq</span></a>,
<a class="reference internal" href="#sdfstats"><span class="std std-ref">sdfstats</span></a></p>
</div>
</div>
<div class="section" id="sdfsubset">
<h3>sdfsubset<a class="headerlink" href="#sdfsubset" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Extracts a specified subset of sequences from one SDF and outputs them
to another SDF.</p>
<p><strong>Syntax:</strong></p>
<p>Individual specification of sequence ids:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfsubset [OPTION]... -i SDF -o SDF STRING+
</pre></div>
</div>
<p>File list specification of sequence ids:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfsubset [OPTION]... -i SDF -o SDF -I FILE
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfsubset -i reads -o subset_reads 10 20 30 40 50
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="16%" />
<col width="33%" />
<col width="51%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-i</span></code></td>
<td><code class="docutils literal"><span class="pre">--input=SDF</span></code></td>
<td>Specifies the input SDF.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-o</span></code></td>
<td><code class="docutils literal"><span class="pre">--output=SDF</span></code></td>
<td>The name of the output SDF.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="13%" />
<col width="81%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Filtering</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--end-id=INT</span></code></td>
<td>Only output sequences with sequence id less than the given number. (Sequence ids start at 0).</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--start-id=INT</span></code></td>
<td>Only output sequences with sequence id greater than or equal to the given number. (Sequence ids start at 0).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-I</span></code></td>
<td><code class="docutils literal"><span class="pre">--id-file=FILE</span></code></td>
<td>Name of a file containing a list of sequences to extract, one per line.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--names</span></code></td>
<td>Interpret any specified sequence as names instead of numeric sequence ids.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">STRING+</span></code></td>
<td>Specifies the sequence id, or sequence name if the names flag is set to extract from the input SDF. May be specified 0 or more times.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="21%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Use this command to obtain a subset of sequences from an SDF. Either
specify the subset on the command line as a list of space-separated
sequence ids or using the <code class="docutils literal"><span class="pre">--id-file</span></code> parameter to specify a file
containing a list of sequence ids, one per line. Sequence ids start from
zero and are the same as the ids that <code class="docutils literal"><span class="pre">map</span></code> uses by default in the
<code class="docutils literal"><span class="pre">QNAME</span></code> field of its BAM files.</p>
<p>For example:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfsubset -i reads -o subset_reads 10 20 30 40 50
</pre></div>
</div>
<p>This will produce an SDF called subset_reads with sequences 10, 20, 30,
40 and 50 from the original SDF contained in it.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#sdfsubseq"><span class="std std-ref">sdfsubseq</span></a>,
<a class="reference internal" href="#sdfstats"><span class="std std-ref">sdfstats</span></a></p>
</div>
</div>
<div class="section" id="sdfsubseq">
<h3>sdfsubseq<a class="headerlink" href="#sdfsubseq" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Prints a subsequence of a given sequence in an SDF.</p>
<p><strong>Syntax:</strong></p>
<p>Print sequences from sequence names:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfsubseq [OPTION]... -i FILE STRING+
</pre></div>
</div>
<p>Print sequences from sequence ids:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfsubseq [OPTION]... -i FILE -I STRING+
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfsubseq -i reads -I 0:1+100
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="17%" />
<col width="34%" />
<col width="48%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-i</span></code></td>
<td><code class="docutils literal"><span class="pre">--input=FILE</span></code></td>
<td>Specifies the input SDF.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="10%" />
<col width="85%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Filtering</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-I</span></code></td>
<td><code class="docutils literal"><span class="pre">--sequence-id</span></code></td>
<td>Set to use sequence id instead of sequence name in region flag (0-based).</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">STRING+</span></code></td>
<td>Specifies the region to display. The format is one of &lt;sequence_name&gt;, &lt;sequence_name&gt;:start-end or &lt;sequence_name&gt;:start+length. Must be specified 1 or more times</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="12%" />
<col width="35%" />
<col width="53%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-f</span></code></td>
<td><code class="docutils literal"><span class="pre">--fasta</span></code></td>
<td>Set to output in FASTA format.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-q</span></code></td>
<td><code class="docutils literal"><span class="pre">--fastq</span></code></td>
<td>Set to output in FASTQ format.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-r</span></code></td>
<td><code class="docutils literal"><span class="pre">--reverse-complement</span></code></td>
<td>Set to output in reverse complement.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Prints out the nucleotides or amino acids of specified regions in a set
of sequences.</p>
<p>For example:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfsubseq --input reads --sequence-id 0:1+20
AGGCGTCTGCAGCCGACGCG
</pre></div>
</div>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#sdfsubset"><span class="std std-ref">sdfsubset</span></a>,
<a class="reference internal" href="#sdfstats"><span class="std std-ref">sdfstats</span></a></p>
</div>
</div>
<div class="section" id="mendelian">
<h3>mendelian<a class="headerlink" href="#mendelian" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>The <code class="docutils literal"><span class="pre">mendelian</span></code> command checks a multi-sample VCF file for variant calls
which do not follow Mendelian inheritance, and compute aggregate sample
concordance.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg mendelian [OPTION]... -i FILE -t SDF
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg mendelian -i family.vcf.gz -t genome_ref
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="24%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-i</span></code></td>
<td><code class="docutils literal"><span class="pre">--input=FILE</span></code></td>
<td>VCF file containing the multiple sample variant calls. Use &#8216;-&#8216; to read from standard input.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--output=FILE</span></code></td>
<td>Set to output annotated calls to this VCF file.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--output-consistent=FILE</span></code></td>
<td>Set to output only consistent calls to this VCF file.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--output-inconsistent=FILE</span></code></td>
<td>Set to output only non-Mendelian calls to this VCF file.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-t</span></code></td>
<td><code class="docutils literal"><span class="pre">--template=SDF</span></code></td>
<td>SDF containing template to which was used to create the VCF.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="18%" />
<col width="76%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Sensitivity Tuning</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-l</span></code></td>
<td><code class="docutils literal"><span class="pre">--lenient</span></code></td>
<td>Set to allow homozygous diploid variant calls in place of haploid calls and assume missing values are equal to the reference.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--all-records</span></code></td>
<td>Use all records, regardless of filters. Default is to only process records where FILTER is &#8221;.&#8221; or &#8220;PASS&#8221;.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--min-concordance=FLOAT</span></code></td>
<td>The percentage concordance required for parentage to be considered as consistent. The default is 99.0.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--pedigree=FILE</span></code></td>
<td>Specify a genome relationships PED file. The default is to extract pedigree information from the VCF header fields.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="11%" />
<col width="83%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-Z</span></code></td>
<td><code class="docutils literal"><span class="pre">--no-gzip</span></code></td>
<td>Set this flag to create the VCF output file without compression. By default the output file is compressed with blocked gzip.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Given a multi-sample VCF file for a nuclear family with a defined
pedigree, the <code class="docutils literal"><span class="pre">mendelian</span></code> command examines the variant calls and outputs
the number of violations of Mendelian inheritance. If the
<code class="docutils literal"><span class="pre">--output-inconsistent</span></code> parameter is set, all detected violations are
written into an output VCF file. As such, this command may be regarded
as a VCF filter, outputting those variant calls needing a non-Mendelian
explanation. Such calls may be the consequence of sequencing error,
calling on low-coverage, or genuine novel variants in one or more
individuals.</p>
<p>Pedigree information regarding the relationships between samples and the
sex of each sample is extracted from the VCF headers automatically
created by the RTG pedigree-aware variant calling commands. If this
pedigree information is absent from the VCF header or is incorrect, a
pedigree file can be explicitly supplied with the <code class="docutils literal"><span class="pre">--pedigree</span></code> flag.</p>
<p>To ensure correct behavior when dealing with sex chromosomes it is
necessary to specify a template and ensure the sex of each sample is
supplied as part of the pedigree information. While it is best to give
the template used in the creation of the VCF, for checking third-party
outputs any template containing the same chromosome names and an
appropriate <code class="docutils literal"><span class="pre">reference.txt</span></code> file will work.</p>
<p>Particularly when evaluating VCF files that have been produced by third
party tools or when the VCF is the result of combining independent
per-sample calling, you can end up with situations where calls are not
available for every member of the family. Under normal circumstances
these will be reported as an allele count constraint violation. It is
possible to treat missing values as equal to the reference by using the
<code class="docutils literal"><span class="pre">--lenient</span></code> parameter. Note that while this approach will be correct in
most cases, it will give inaccurate results where the calling between
different samples has reported the variant in an equivalent but slightly
different position or representation (e.g. positioning of indels within
homopolymer regions, differences of representation such as splitting
MNPs into multiple SNPs etc).</p>
<p>The <code class="docutils literal"><span class="pre">mendelian</span></code> command computes overall concordance between related
samples to assist detecting cases where pedigree has been incorrectly
recorded or samples have been mislabelled. For each child in the
pedigree, pairwise concordance is computed with respect to each parent
by identifying diploid calls where the parent does not contain either
allele called in the child. Low pairwise concordance with a single
parent may indicate that the parent is the source of the problem,
whereas low pairwise concordance with both parents may indicate that the
child is the source of the problem. A stricter three-way concordance is
also recorded.</p>
<p>By default, only VCF records with the FILTER field set to PASS or
missing are processed. All variant records can be examined by specifying
the <code class="docutils literal"><span class="pre">--all-records</span></code> parameter.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#family"><span class="std std-ref">family</span></a>,
<a class="reference internal" href="#population"><span class="std std-ref">population</span></a>,
<a class="reference internal" href="#vcfstats"><span class="std std-ref">vcfstats</span></a></p>
</div>
</div>
<div class="section" id="vcfstats">
<h3>vcfstats<a class="headerlink" href="#vcfstats" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Display simple statistics about the contents of a set of VCF files.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfstats [OPTION]... FILE+
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfstats /data/human/wgs/NA19240/snp_chr5.vcf.gz

Location                      : /data/human/wgs/NA19240/snp_chr5.vcf.gz
Passed Filters                : 283144
Failed Filters                : 83568
SNPs                          : 241595
MNPs                          : 5654
Insertions                    : 15424
Deletions                     : 14667
Indels                        : 1477
Unchanged                     : 4327
SNP Transitions/Transversions : 1.93 (210572/108835)
Total Het/Hom ratio           : 2.13 (189645/89172)
SNP Het/Hom ratio             : 2.12 (164111/77484)
MNP Het/Hom ratio             : 3.72 (4457/1197)
Insertion Het/Hom ratio       : 1.69 (9695/5729)
Deletion Het/Hom ratio        : 2.33 (10263/4404)
Indel Het/Hom ratio           : 3.13 (1119/358)
Insertion/Deletion ratio      : 1.05 (15424/14667)
Indel/SNP+MNP ratio           : 0.13 (31568/247249)
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="3%" />
<col width="13%" />
<col width="84%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--known</span></code></td>
<td>Set to only calculate statistics for known variants.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--novel</span></code></td>
<td>Set to only calculate statistics for novel variants.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--sample=FILE</span></code></td>
<td>Set to only calculate statistics for the specified sample. (Default is to include all samples). May be specified 0 or more times.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td>FILE+</td>
<td>VCF file from which to derive statistics. Use &#8216;-&#8216; to read from standard input. Must be specified 1 or more times.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="34%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Reporting</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--allele-lengths</span></code></td>
<td>Set to output variant length histogram.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="21%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Use the <code class="docutils literal"><span class="pre">vcfstats</span></code> command to display summary statistics for a set of
VCF files. If a VCF file contains multiple sample columns, the
statistics for each sample are shown individually.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#snp"><span class="std std-ref">snp</span></a>,
<a class="reference internal" href="#family"><span class="std std-ref">family</span></a>,
<a class="reference internal" href="#somatic"><span class="std std-ref">somatic</span></a>,
<a class="reference internal" href="#vcfmerge"><span class="std std-ref">vcfmerge</span></a>,
<a class="reference internal" href="#discord"><span class="std std-ref">discord</span></a></p>
</div>
</div>
<div class="section" id="vcfmerge">
<h3>vcfmerge<a class="headerlink" href="#vcfmerge" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Combines the contents of two or more VCF files. The <code class="docutils literal"><span class="pre">vcfmerge</span></code> command
can concatenate the outputs of per-chromosome variant detection runs to
create a complete genome VCF file, and also merge VCF outputs from
multiple samples to form a multi-sample VCF file.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfmerge [OPTION]... -o FILE FILE+
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfmerge -o merged.vcf.gz snp1.vcf.gz snp2.vcf.gz
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="22%" />
<col width="70%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-a</span></code></td>
<td><code class="docutils literal"><span class="pre">--add-header=STRING</span></code></td>
<td>Add the supplied text to the output VCF header. May be specified 0 or more times.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-o</span></code></td>
<td><code class="docutils literal"><span class="pre">--output=FILE</span></code></td>
<td>The output VCF file name. Use &#8216;-&#8216; to write to standard output.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td>FILE+</td>
<td>VCF files to be merged. Must be specified 1 or more times.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="18%" />
<col width="76%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-f</span></code></td>
<td><code class="docutils literal"><span class="pre">--force-merge=STRING</span></code></td>
<td>Set to allow merging of specified header ID even when descriptions do not match. May be specified 0 or more times.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-F</span></code></td>
<td><code class="docutils literal"><span class="pre">--force-merge-all</span></code></td>
<td>Attempt merging of all non-matching header declarations.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-Z</span></code></td>
<td><code class="docutils literal"><span class="pre">--no-gzip</span></code></td>
<td>Set this flag to create the VCF output file without compression. By default the output file is compressed with
blocked gzip.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--no-index</span></code></td>
<td>Set this flag to not produce the index for the VCF output file.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--preserve-formats</span></code></td>
<td>If set, variants with different ALTs and unmergeable FORMAT fields will be kept unmerged (Default is to remove those
FORMAT fields so the variants can be combined).</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--stats</span></code></td>
<td>Set to output statistics for the merged VCF file.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>The <code class="docutils literal"><span class="pre">vcfmerge</span></code> command takes a list of VCF files and outputs to a single
VCF file. The input files must have consistent header lines, although
similar header lines can be forced to merge using the <code class="docutils literal"><span class="pre">--force-merge</span></code>
parameter. Each VCF file must be block compressed and have a
corresponding tabix index file, which is the default for outputs from
RTG variant detection tools, but may also be created from an existing
VCF file using the RTG <code class="docutils literal"><span class="pre">bgzip</span></code> and <code class="docutils literal"><span class="pre">index</span></code> commands.</p>
<p>There are two primary usage scenarios for the <code class="docutils literal"><span class="pre">vcfmerge</span></code> command. The
first is to combine input VCFs corresponding to different genomic
regions (for example, if variant calling was carried out for each
chromosome independently on different nodes of a compute cluster). The
second scenario is when combining VCFs containing variant calls for
different samples (e.g. combining calls made for separate cohorts into a
single VCF). If the input VCFs contain multiple calls at the same
position for the same sample, a warning is issued and only the first is
kept.</p>
<p>When multiple records occur at the same position and the length on the
reference is the same, the records will be merged into a single record.
If the merge results in a change in the set of ALT alleles, any VCF
<code class="docutils literal"><span class="pre">FORMAT</span></code> fields declared to be of type <code class="docutils literal"><span class="pre">A</span></code>, <code class="docutils literal"><span class="pre">G</span></code>, or <code class="docutils literal"><span class="pre">R</span></code> will be
set to the missing value (<code class="docutils literal"><span class="pre">.</span></code>), as they cannot be meaningfully updated.
The <code class="docutils literal"><span class="pre">--preserve-formats</span></code> flag prevents this loss of information by
refusing to merge the records (separate records will be output).</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#snp"><span class="std std-ref">snp</span></a>,
<a class="reference internal" href="#family"><span class="std std-ref">family</span></a>,
<a class="reference internal" href="#population"><span class="std std-ref">population</span></a>,
<a class="reference internal" href="#somatic"><span class="std std-ref">somatic</span></a>,
<a class="reference internal" href="#discord"><span class="std std-ref">discord</span></a>,
<a class="reference internal" href="#bgzip"><span class="std std-ref">bgzip</span></a>,
:ref: <cite>index</cite></p>
</div>
</div>
<div class="section" id="vcffilter">
<h3>vcffilter<a class="headerlink" href="#vcffilter" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Filter VCF output files to include or exclude records based on various
criteria.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcffilter [OPTION]... -i FILE -o FILE
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcffilter -i snps.vcf.gz -o snps_cov5.vcf.gz -d 5
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="15%" />
<col width="79%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--all-samples</span></code></td>
<td>Set to apply sample-specific criteria to all samples contained in the input VCF.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--bed-regions=FILE</span></code></td>
<td>If set, only read VCF records that overlap the ranges contained in the specified BED file. Requires the input VCF to be tabix indexed.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-i</span></code></td>
<td><code class="docutils literal"><span class="pre">--input=FILE</span></code></td>
<td>Specifies the VCF file containing variants to be filtered. Use &#8216;-&#8216; to read from standard input.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-o</span></code></td>
<td><code class="docutils literal"><span class="pre">--output=FILE</span></code></td>
<td>Specifies the output VCF file. Use &#8216;-&#8216; to write to standard output. This option is required, unless <code class="docutils literal"><span class="pre">--javascript</span></code> is being used.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--region=STRING</span></code></td>
<td>if set, only read VCF records within the specified range. The format is one of &lt;template_name&gt;, &lt;template_name&gt;:start-end.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--sample=STRING</span></code></td>
<td>Set to apply sample-specific criteria to the named sample contained in the input VCF. May be specified 0 or more times.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="28%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Filtering (Record based)</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-w</span></code></td>
<td><code class="docutils literal"><span class="pre">--density-window=INT</span></code></td>
<td>Set a window length in which multiple called variants are discarded.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--exclude-bed=FILE</span></code></td>
<td>Set to discard all variants within the regions contained in the BED file.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--exclude-vcf=FILE</span></code></td>
<td>Set to discard all variants that overlap with the ones in this VCF file.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--include-bed=FILE</span></code></td>
<td>Set to only keep variants within the regions contained in the BED file.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--include-vcf=FILE</span></code></td>
<td>Set to only keep variants that overlap with the ones in this VCF file.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-j</span></code></td>
<td><code class="docutils literal"><span class="pre">--javascript=STRING</span></code></td>
<td>Specify filtering functions in javascript. May be either an expression
or a file name. May be specified 0 or more times.
<a class="reference internal" href="#javascript-examples"><span class="std std-ref">See Examples</span></a></td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-e</span></code></td>
<td><code class="docutils literal"><span class="pre">--keep-expr=STRING</span></code></td>
<td>Set to only keep variants for which this expression evaluates to true.
<a class="reference internal" href="#expression-examples"><span class="std std-ref">See Examples</span></a></td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-k</span></code></td>
<td><code class="docutils literal"><span class="pre">--keep-filter=STRING</span></code></td>
<td>Set to only keep variants with this FILTER tag. May be specified 0 or more times,
or as a comma separated list.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-K</span></code></td>
<td><code class="docutils literal"><span class="pre">--keep-info=STRING</span></code></td>
<td>Set to only keep variants with this INFO tag. May be specified 0 or more times,
or as a comma separated list.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-C</span></code></td>
<td><code class="docutils literal"><span class="pre">--max-combined-read-depth=INT</span></code></td>
<td>Set the maximum allowed combined read depth.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-Q</span></code></td>
<td><code class="docutils literal"><span class="pre">--max-quality=FLOAT</span></code></td>
<td>Set the maximum allowed quality.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-c</span></code></td>
<td><code class="docutils literal"><span class="pre">--min-combined-read-depth=INT</span></code></td>
<td>Set the minimum allowed combined read depth.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-q</span></code></td>
<td><code class="docutils literal"><span class="pre">--min-quality=FLOAT</span></code></td>
<td>Set the minimum allowed quality.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--non-snps-only</span></code></td>
<td>Set to output MNPs and INDELs only.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-r</span></code></td>
<td><code class="docutils literal"><span class="pre">--remove-filter=STRING</span></code></td>
<td>Set to remove variants with this FILTER tag. May be specified 0 or more times,
or as a comma separated list.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-R</span></code></td>
<td><code class="docutils literal"><span class="pre">--remove-info=STRING</span></code></td>
<td>Set to remove variants with this INFO tag. May be specified 0 or more times,
or as a comma separated list.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--remove-overlapping</span></code></td>
<td>Set to remove records that overlap with previous records.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--snps-only</span></code></td>
<td>Set to output simple SNPs only.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="28%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Filtering (Sample based)</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-A</span></code></td>
<td><code class="docutils literal"><span class="pre">--max-ambiguity-ratio=FLOAT</span></code></td>
<td>Set the maximum allowed ambiguity ratio.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--max-avr-score=FLOAT</span></code></td>
<td>Set the maximum allowed AVR score.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--max-denovo-score=FLOAT</span></code></td>
<td>Set the maximum allowed de novo score.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-G</span></code></td>
<td><code class="docutils literal"><span class="pre">--max-genotype-quality=FLOAT</span></code></td>
<td>Set the maximum allowed genotype quality.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-D</span></code></td>
<td><code class="docutils literal"><span class="pre">--max-read-depth=INT</span></code></td>
<td>Set the maximum allowed sample read depth.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--min-avr-score=FLOAT</span></code></td>
<td>Set the minimum allowed AVR score.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--min-denovo-score=FLOAT</span></code></td>
<td>Set the minimum allowed de novo score.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-g</span></code></td>
<td><code class="docutils literal"><span class="pre">--min-genotype-quality=FLOAT</span></code></td>
<td>Set the minimum allowed genotype quality.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-d</span></code></td>
<td><code class="docutils literal"><span class="pre">--min-read-depth=INT</span></code></td>
<td>Set the minimum allowed sample read depth.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--remove-all-same-as-ref</span></code></td>
<td>Set to remove records where all the samples are same as the reference.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--remove-hom</span></code></td>
<td>Remove where sample is homozygous.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--remove-same-as-ref</span></code></td>
<td>Set to remove variants where the sample is the same as reference.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="3%" />
<col width="22%" />
<col width="75%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Reporting</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--clear-failed-samples</span></code></td>
<td>Set to have the GT field of failing samples set to the missing value instead of removing the record.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--fail=STRING</span></code></td>
<td>Set to have the filter field of a failed record set to the provided value instead of removing it.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="11%" />
<col width="83%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-Z</span></code></td>
<td><code class="docutils literal"><span class="pre">--no-gzip</span></code></td>
<td>Set this flag to create the output file without compression. By default the output file is compressed with tabix compatible blocked gzip.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--no-header</span></code></td>
<td>prevent VCF header from being written</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--no-index</span></code></td>
<td>Set this flag to not produce the tabix index for the output file.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Use <code class="docutils literal"><span class="pre">vcffilter</span></code> to get a subset of the results from variant calling
based on the filtering criteria supplied by the filter flags. When
filtering on multiple samples, if any of the specified samples fail the
criteria, the record will be filtered. The default behavior is for
filtered records to be excluded from output altogether, but
alternatively the records can be retained but with an additional
user-specified VCF FILTER status set via <code class="docutils literal"><span class="pre">--fail</span></code> option, or if
sample-specific filtering criteria is being applied, only those samples
can be filtered by setting their GT field to missing by using the
<code class="docutils literal"><span class="pre">--clear-failed-samples</span></code> option.</p>
<p>The <code class="docutils literal"><span class="pre">--bed-regions</span></code> option makes use of tabix indexes to avoid loading
VCF records outside the supplied regions, which can give faster
filtering performance. If the input VCF is not indexed or being read
from standard input, or if records failing filters are to be annotated
via the <code class="docutils literal"><span class="pre">--fail</span></code> option, use the <code class="docutils literal"><span class="pre">--include-bed</span></code> option instead.</p>
<p>The flags <code class="docutils literal"><span class="pre">--min-denovo-score</span></code> and <code class="docutils literal"><span class="pre">--max-denovo-score</span></code> can only be used
on a single sample. Records will only be kept if the specified sample is
flagged as a <em>de novo</em> variant and the score is within the range
specified by the flags. It will also only be kept if none of the other
samples for the record are also flagged as a <em>de novo</em> variant within
the specified score range.</p>
<p>A powerful general-purpose filtering capability has been included that
permits the specification of filter criteria as simple JavaScript
expressions (<code class="docutils literal"><span class="pre">--keep-expr</span></code>) or more comprehensive JavaScript processing
functions (<code class="docutils literal"><span class="pre">--javascript</span></code>). Both <code class="docutils literal"><span class="pre">--keep-expr</span></code> and <code class="docutils literal"><span class="pre">--javascript</span></code> can
take JavaScript on the command line or if a filename is supplied then
the script/expression will be read from that file. <code class="docutils literal"><span class="pre">--keep-expr</span></code> will be
applied before <code class="docutils literal"><span class="pre">--javascript</span></code>, so the <code class="docutils literal"><span class="pre">--javascript</span></code> record
function will not be called for records filtered out by <code class="docutils literal"><span class="pre">--keep-expr</span></code>.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For full details of functions available in <code class="docutils literal"><span class="pre">--keep-expr</span></code> and <code class="docutils literal"><span class="pre">--javascript</span></code> see <a class="reference internal" href="appendix.html#rtg-javascript-filtering-api"><span class="std std-ref">RTG JavaScript filtering API</span></a></p>
</div>
<div class="section" id="simple-filtering-by-javascript-expression-with-keep-expr">
<h4>Simple filtering by JavaScript expression with <code class="docutils literal"><span class="pre">--keep-expr</span></code><a class="headerlink" href="#simple-filtering-by-javascript-expression-with-keep-expr" title="Permalink to this headline">¶</a></h4>
<p>The <code class="docutils literal"><span class="pre">--keep-expr</span></code> flag aims to provide a convenient way to apply some
simple (typically one line) filtering expressions which are evaluated in
the context of each record. The final expression of the fragment must
evaluate to a boolean value. Records which evaluate to <code class="docutils literal"><span class="pre">true</span></code> will be
retained, while <code class="docutils literal"><span class="pre">false</span></code> will be removed. The value must be of type
boolean, simply being truthy/falsy (in the JavaScript sense) will raise
an error.</p>
<div class="section" id="keep-expr-examples">
<span id="expression-examples"></span><h5><code class="docutils literal"><span class="pre">--keep-expr</span></code> examples:<a class="headerlink" href="#keep-expr-examples" title="Permalink to this headline">¶</a></h5>
<p>The following expression keeps records where the <code class="docutils literal"><span class="pre">NA12878</span></code> sample has
<code class="docutils literal"><span class="pre">GQ</span></code> &gt; 30 and the total depth is &gt; 20. JavaScript will auto convert
numerical strings when comparing a string with a number, so calls to parseInt can be omitted.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcffilter -i in.vcf.gz -o out.vcf.gz \
--keep-expr &quot;&#39;NA12878&#39;.GQ &gt; 30 &amp;&amp; INFO.DP &gt; 20&quot;
</pre></div>
</div>
<p>If the field of interest may contain the missing value (&#8216;.&#8217;) or may be
entirely missing on a per-record basis, the <code class="docutils literal"><span class="pre">has()</span></code> function can be
used to control whether such records are kept vs filtered. For example,
to keep records with depth greater than 20, and remove any without a DP
annotation:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcffilter -i in.vcf.gz -o out.vcf.gz \
--keep-expr &quot;has(INFO.DP) &amp;&amp; INFO.DP &gt; 20&quot;
</pre></div>
</div>
<p>Alternatively, to keep records with depth greater than 20, as well as those without
a DP annotation:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcffilter -i in.vcf.gz -o out.vcf.gz \
--keep-expr &quot;!has(INFO.DP) || INFO.DP &gt; 20&quot;
</pre></div>
</div>
<p>The next example keeps records where all samples have a depth &gt; 10. The
standard JavaScript array methods <code class="docutils literal"><span class="pre">every</span></code> and <code class="docutils literal"><span class="pre">some</span></code> can be used to
apply a condition on every sample column.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcffilter -i in.vcf.gz -o out.vcf.gz \
  --keep-expr &quot;SAMPLES.every(function(s) {return s.DP &gt; 10})&quot;
</pre></div>
</div>
</div>
</div>
<div class="section" id="advanced-javascript-filtering-with-javascript">
<h4>Advanced JavaScript filtering with <code class="docutils literal"><span class="pre">--javascript</span></code><a class="headerlink" href="#advanced-javascript-filtering-with-javascript" title="Permalink to this headline">¶</a></h4>
<p>The <code class="docutils literal"><span class="pre">--javascript</span></code> option aims to support more complicated processing
than <code class="docutils literal"><span class="pre">--keep-expr</span></code>. permitting modification of the output VCF, or
supporting use cases where the script is tasked to compute and output
alternative information in addition to (or instead of) the output VCF.
The scripts specified by the user are evaluated once at the start of
processing. Two special functions may be defined in a <code class="docutils literal"><span class="pre">--javascript</span></code>
script, which will then be executed in different contexts:</p>
<ul class="simple">
<li>A function with the name <code class="docutils literal"><span class="pre">record</span></code> will be executed once for each VCF
record. If the <code class="docutils literal"><span class="pre">record</span></code> function has a return value it must have type
boolean. Records which evaluate to <code class="docutils literal"><span class="pre">true</span></code> will be retained, while
<code class="docutils literal"><span class="pre">false</span></code> will be removed. If the record function has no return value
then the record will be retained. The <code class="docutils literal"><span class="pre">record</span></code> function is applied
after any <code class="docutils literal"><span class="pre">--keep-expr</span></code> expression.</li>
<li>A function with the name <code class="docutils literal"><span class="pre">end</span></code> will be called once at the end of
processing. This allows reporting of summary statistics collected
during the filter process.</li>
</ul>
<p>This <code class="docutils literal"><span class="pre">--javascript</span></code> flag may be specified multiple times, they will be
evaluated in order, in a shared JavaScript namespace, before VCF
processing commences. This permits a use case where an initial
JavaScript expression supplies parameter values which will be required
by a subsequent JavaScript file.</p>
<div class="section" id="example-javascript-scripts">
<span id="javascript-examples"></span><h5>Example <code class="docutils literal"><span class="pre">--javascript</span></code> scripts:<a class="headerlink" href="#example-javascript-scripts" title="Permalink to this headline">¶</a></h5>
<p>To find indels with length greater than 5, save the following to a file
named <code class="docutils literal"><span class="pre">find-indels.js</span></code>:</p>
<div class="highlight-javascript"><div class="highlight"><pre><span></span><span class="c1">// Finds indels with length &gt; 5</span>
<span class="kd">function</span> <span class="nx">record</span><span class="p">()</span> <span class="p">{</span>
  <span class="kd">var</span> <span class="nx">deltas</span> <span class="o">=</span> <span class="nx">ALT</span><span class="p">.</span><span class="nx">map</span><span class="p">(</span><span class="kd">function</span> <span class="p">(</span><span class="nx">alt</span><span class="p">)</span> <span class="p">{</span>
    <span class="k">return</span> <span class="nb">Math</span><span class="p">.</span><span class="nx">abs</span><span class="p">(</span><span class="nx">alt</span><span class="p">.</span><span class="nx">length</span> <span class="o">-</span> <span class="nx">REF</span><span class="p">.</span><span class="nx">length</span><span class="p">);</span>
  <span class="p">});</span>
  <span class="k">return</span> <span class="nx">deltas</span><span class="p">.</span><span class="nx">some</span><span class="p">(</span><span class="kd">function</span> <span class="p">(</span><span class="nx">delta</span><span class="p">)</span> <span class="p">{</span><span class="k">return</span> <span class="nx">delta</span> <span class="o">&gt;</span> <span class="mi">5</span><span class="p">});</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Then perform the filtering via:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcffilter -i in.vcf.gz -o out.vcf.gz --javascript find-indels.js
</pre></div>
</div>
<p>The following example derives a new FORMAT column containing variant
allelic fraction based on the values in the AD and DP FORMAT
annotations, for every sample contained in the VCF. Save the following
to a file named <code class="docutils literal"><span class="pre">add-vaf.js</span></code>:</p>
<div class="highlight-javascript"><div class="highlight"><pre><span></span><span class="c1">// Derive new VAF FORMAT field for each sample</span>
<span class="nx">ensureFormatHeader</span><span class="p">(</span><span class="s1">&#39;##FORMAT=&lt;ID=VAF,Number=1,Type=Float,&#39;</span> <span class="o">+</span>
  <span class="s1">&#39;Description=&quot;Variant Allelic Fraction&quot;&gt;&#39;</span><span class="p">);</span>

<span class="kd">function</span> <span class="nx">record</span><span class="p">()</span> <span class="p">{</span>
  <span class="nx">SAMPLES</span><span class="p">.</span><span class="nx">forEach</span><span class="p">(</span><span class="kd">function</span><span class="p">(</span><span class="nx">sample</span><span class="p">)</span> <span class="p">{</span>
    <span class="c1">// Take all but the first AD value as numerics</span>
    <span class="kd">var</span> <span class="nx">altDepths</span> <span class="o">=</span> <span class="nx">sample</span><span class="p">.</span><span class="nx">AD</span><span class="p">.</span><span class="nx">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">).</span><span class="nx">slice</span><span class="p">(</span><span class="mi">1</span><span class="p">);</span>
    <span class="c1">// Find the max</span>
    <span class="kd">var</span> <span class="nx">maxAltDepth</span> <span class="o">=</span> <span class="nb">Math</span><span class="p">.</span><span class="nx">max</span><span class="p">.</span><span class="nx">apply</span><span class="p">(</span><span class="kc">null</span><span class="p">,</span> <span class="nx">altDepths</span><span class="p">);</span>
    <span class="k">if</span> <span class="p">(</span><span class="nx">maxAltDepth</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
      <span class="nx">sample</span><span class="p">.</span><span class="nx">VAF</span> <span class="o">=</span> <span class="nx">sample</span><span class="p">.</span><span class="nx">DP</span> <span class="o">/</span> <span class="nx">maxAltDepth</span><span class="p">;</span>
    <span class="p">}</span>
  <span class="p">});</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Then run the filtering via:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcffilter -i in.vcf.gz -o out.vcf.gz --javascript add-vaf.js
</pre></div>
</div>
<p>The next example produces a table of binned indel lengths, save the
following to a file named <code class="docutils literal"><span class="pre">indel-lengths.js</span></code>:</p>
<div class="highlight-javascript"><div class="highlight"><pre><span></span><span class="c1">// bin breakpoints can be customised by defining your own bins[] in a</span>
<span class="c1">// previous -j flag</span>
<span class="k">if</span> <span class="p">(</span><span class="k">typeof</span> <span class="nx">bins</span> <span class="o">==</span> <span class="s2">&quot;undefined&quot;</span><span class="p">)</span> <span class="p">{</span>
  <span class="kd">var</span> <span class="nx">bins</span> <span class="o">=</span> <span class="p">[</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="o">-</span><span class="mi">5</span><span class="p">,</span> <span class="o">-</span><span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">11</span><span class="p">];</span>
<span class="p">}</span>

<span class="kd">var</span> <span class="nx">counts</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">];</span>
<span class="nx">bins</span><span class="p">.</span><span class="nx">forEach</span><span class="p">(</span><span class="kd">function</span> <span class="p">()</span> <span class="p">{</span><span class="nx">counts</span><span class="p">.</span><span class="nx">push</span><span class="p">(</span><span class="mi">0</span><span class="p">)});</span>
<span class="kd">function</span> <span class="nx">record</span><span class="p">()</span> <span class="p">{</span>
  <span class="k">if</span> <span class="p">(</span><span class="nx">ALT</span><span class="p">.</span><span class="nx">length</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
    <span class="k">return</span> <span class="kc">false</span><span class="p">;</span>
  <span class="p">}</span>
  <span class="kd">var</span> <span class="nx">deltas</span> <span class="o">=</span> <span class="nx">ALT</span><span class="p">.</span><span class="nx">map</span><span class="p">(</span><span class="kd">function</span> <span class="p">(</span><span class="nx">alt</span><span class="p">)</span> <span class="p">{</span> <span class="k">return</span> <span class="nx">alt</span><span class="p">.</span><span class="nx">length</span> <span class="o">-</span> <span class="nx">REF</span><span class="p">.</span><span class="nx">length</span><span class="p">;</span> <span class="p">});</span>
  <span class="kd">var</span> <span class="nx">maxDel</span> <span class="o">=</span> <span class="nb">Math</span><span class="p">.</span><span class="nx">min</span><span class="p">.</span><span class="nx">apply</span><span class="p">(</span><span class="kc">null</span><span class="p">,</span> <span class="nx">deltas</span><span class="p">);</span>
  <span class="kd">var</span> <span class="nx">maxIns</span> <span class="o">=</span> <span class="nb">Math</span><span class="p">.</span><span class="nx">max</span><span class="p">.</span><span class="nx">apply</span><span class="p">(</span><span class="kc">null</span><span class="p">,</span> <span class="nx">deltas</span><span class="p">);</span>
  <span class="kd">var</span> <span class="nx">delta</span> <span class="o">=</span> <span class="nb">Math</span><span class="p">.</span><span class="nx">abs</span><span class="p">(</span><span class="nx">maxDel</span><span class="p">)</span> <span class="o">&gt;</span> <span class="nx">maxIns</span> <span class="o">?</span> <span class="nx">maxDel</span> <span class="o">:</span> <span class="nx">maxIns</span><span class="p">;</span>

  <span class="k">if</span> <span class="p">(</span><span class="nx">delta</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
    <span class="k">return</span> <span class="kc">false</span><span class="p">;</span>
  <span class="p">}</span>
  <span class="k">for</span> <span class="p">(</span><span class="kd">var</span> <span class="nx">i</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span> <span class="nx">i</span> <span class="o">&lt;</span> <span class="nx">bins</span><span class="p">.</span><span class="nx">length</span><span class="p">;</span> <span class="nx">i</span><span class="o">++</span><span class="p">)</span> <span class="p">{</span>
    <span class="k">if</span> <span class="p">(</span><span class="nx">delta</span> <span class="o">&lt;</span> <span class="nx">bins</span><span class="p">[</span><span class="nx">i</span><span class="p">])</span> <span class="p">{</span>
      <span class="nx">counts</span><span class="p">[</span><span class="nx">i</span><span class="p">]</span><span class="o">++</span><span class="p">;</span>
      <span class="k">break</span><span class="p">;</span>
    <span class="p">}</span>
  <span class="p">}</span>
  <span class="k">if</span> <span class="p">(</span><span class="nx">delta</span> <span class="o">&gt;</span> <span class="nx">bins</span><span class="p">[</span><span class="nx">bins</span><span class="p">.</span><span class="nx">length</span> <span class="o">-</span> <span class="mi">1</span><span class="p">])</span> <span class="p">{</span>
    <span class="nx">counts</span><span class="p">[</span><span class="nx">counts</span><span class="p">.</span><span class="nx">length</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">++</span><span class="p">;</span>
  <span class="p">}</span>
  <span class="k">return</span> <span class="kc">false</span><span class="p">;</span>
<span class="p">}</span>

<span class="kd">function</span> <span class="nx">end</span><span class="p">()</span> <span class="p">{</span>
  <span class="nx">print</span><span class="p">(</span><span class="s2">&quot;Delta\\tCount&quot;</span><span class="p">);</span>
  <span class="k">for</span> <span class="p">(</span><span class="kd">var</span> <span class="nx">i</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span> <span class="nx">i</span> <span class="o">&lt;</span> <span class="nx">bins</span><span class="p">.</span><span class="nx">length</span><span class="p">;</span> <span class="nx">i</span><span class="o">++</span><span class="p">)</span> <span class="p">{</span>
    <span class="nx">print</span><span class="p">(</span><span class="s2">&quot;&lt;&quot;</span> <span class="o">+</span> <span class="nx">bins</span><span class="p">[</span><span class="nx">i</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;\\t&quot;</span> <span class="o">+</span> <span class="nx">counts</span><span class="p">[</span><span class="nx">i</span><span class="p">]);</span>
  <span class="p">}</span>
  <span class="nx">print</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span> <span class="o">+</span> <span class="nx">bins</span><span class="p">[</span><span class="nx">bins</span><span class="p">.</span><span class="nx">length</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;\\t&quot;</span> <span class="o">+</span> <span class="nx">counts</span><span class="p">[</span><span class="nx">counts</span><span class="p">.</span><span class="nx">length</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]);</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Then run the filtering via:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcffilter -i in.vcf.gz -o out.vcf.gz --javascript indel-lengths.js
</pre></div>
</div>
<p>We could use this same script with adjusted bins and omitting the output
of the VCF via:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcffilter -i in.vcf.gz -j &quot;var bins = [-20, -10, 0, 20, 20];&quot; \
  -j indel-lengths.js
</pre></div>
</div>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#snp"><span class="std std-ref">snp</span></a>,
<a class="reference internal" href="#family"><span class="std std-ref">family</span></a>,
<a class="reference internal" href="#somatic"><span class="std std-ref">somatic</span></a>,
<a class="reference internal" href="#population"><span class="std std-ref">population</span></a>,
<a class="reference internal" href="#vcfannotate"><span class="std std-ref">vcfannotate</span></a>,
<a class="reference internal" href="#vcfsubset"><span class="std std-ref">vcfsubset</span></a></p>
</div>
</div>
</div>
</div>
<div class="section" id="vcfannotate">
<h3>vcfannotate<a class="headerlink" href="#vcfannotate" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Used to add annotations to a VCF file, either to the VCF <code class="docutils literal"><span class="pre">ID</span></code> field, or
as a VCF <code class="docutils literal"><span class="pre">INFO</span></code> sub-field.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfannotate [OPTION]... -b FILE -i FILE -o FILE
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfannotate -b dbsnp.bed -i snps.vcf.gz -o snps-dbsnp.vcf.gz
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="16%" />
<col width="76%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-i</span></code></td>
<td><code class="docutils literal"><span class="pre">--input=FILE</span></code></td>
<td>Specifies the VCF file containing variants to annotate. Use &#8216;-&#8216; to read from standard input.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-o</span></code></td>
<td><code class="docutils literal"><span class="pre">--output=FILE</span></code></td>
<td>Specifies the output VCF file for the annotated variants. Use &#8216;-&#8216; to write to standard output.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="3%" />
<col width="21%" />
<col width="76%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Reporting</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--bed-ids=FILE</span></code></td>
<td>Specifies a file in BED format containing variant ids in the name column to be added to the VCF id field. May be
specified 0 or more times.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--bed-info=FILE</span></code></td>
<td>Specifies a file in BED format containing annotations in the name column to be added to the VCF info field. May be
specified 0 or more times.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--fill-an-ac</span></code></td>
<td>Set to add or update the AN and AC info fields to the VCF.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--info-description=STRING</span></code></td>
<td>If the BED INFO field is not already declared, use this description in the header. (Default is &#8216;Annotation&#8217;).</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--info-id=STRING</span></code></td>
<td>The INFO ID for BED INFO annotations. (Default is &#8216;ANN&#8217;)</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--relabel</span></code></td>
<td>Relabel samples according to <code class="docutils literal"><span class="pre">old-name&nbsp;new-name</span></code> pairs in specified file. If only a single sample needs to be
relabelled then a construct like <code class="docutils literal"><span class="pre">&lt;(echo</span> <span class="pre">old-name</span> <span class="pre">new-name)</span></code> can be used.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--vcf-ids=FILE</span></code></td>
<td>Specifies a file in VCF format containing variant ids to be added to the VCF id field. May be specified 0 or more
times.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="11%" />
<col width="83%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-Z</span></code></td>
<td><code class="docutils literal"><span class="pre">--no-gzip</span></code></td>
<td>Set this flag to create the output file without compression. By default the output file is compressed with tabix compatible blocked gzip.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--no-index</span></code></td>
<td>Set this flag to not produce the tabix index for the output file.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Use <code class="docutils literal"><span class="pre">vcfannotate</span></code> to add text annotations to variants that fall within
ranges specified in a BED file. The annotations from the BED file are
added as an <code class="docutils literal"><span class="pre">INFO</span></code> field in the output VCF file.</p>
<p>If the <code class="docutils literal"><span class="pre">--bed-ids</span></code> flag is used, instead of adding the annotation to the
<code class="docutils literal"><span class="pre">INFO</span></code> fields, it is added to the <code class="docutils literal"><span class="pre">ID</span></code> column of the VCF file instead.
If the <code class="docutils literal"><span class="pre">--vcf-ids</span></code> flag is used, the <code class="docutils literal"><span class="pre">ID</span></code> column of the input VCF file
is used to update the <code class="docutils literal"><span class="pre">ID</span></code> column of the output VCF file instead.</p>
<p>If the <code class="docutils literal"><span class="pre">--fill-an-ac</span></code> flag is set, the output VCF will have the <code class="docutils literal"><span class="pre">AN</span></code> and
<code class="docutils literal"><span class="pre">AC</span></code> info fields (as defined in the VCF 4.1 specification) created or
updated.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#snp"><span class="std std-ref">snp</span></a>,
<a class="reference internal" href="#family"><span class="std std-ref">family</span></a>,
<a class="reference internal" href="#somatic"><span class="std std-ref">somatic</span></a>,
<a class="reference internal" href="#population"><span class="std std-ref">population</span></a>,
<a class="reference internal" href="#vcffilter"><span class="std std-ref">vcffilter</span></a>,
<a class="reference internal" href="#vcfsubset"><span class="std std-ref">vcfsubset</span></a></p>
</div>
</div>
<div class="section" id="vcfsubset">
<h3>vcfsubset<a class="headerlink" href="#vcfsubset" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Create a VCF file containing a subset of the original columns.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfsubset [OPTION]... -i FILE -o FILE
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfsubset -i snps.vcf.gz -o frequency.vcf.gz --keep-info AF --remove-samples
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="16%" />
<col width="76%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-i</span></code></td>
<td><code class="docutils literal"><span class="pre">--input=FILE</span></code></td>
<td>Specifies the VCF file containing variants to manipulate. Use &#8216;-&#8216; to read from standard input.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-o</span></code></td>
<td><code class="docutils literal"><span class="pre">--output=FILE</span></code></td>
<td>Specifies the output VCF file for the subset records. Use &#8216;-&#8216; to write to standard output.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="3%" />
<col width="24%" />
<col width="73%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Filtering</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--keep-filter=STRING</span></code></td>
<td>Specifies a VCF FILTER tag to keep in the output. May be specified 0 or more times.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--keep-format=STRING</span></code></td>
<td>Specifies a VCF FORMAT tag to keep in the output. May be specified 0 or more times.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--keep-info=STRING</span></code></td>
<td>Specifies a VCF INFO tag to keep in the output. May be specified 0 or more times.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--keep-sample=STRING</span></code></td>
<td>Specifies a sample to keep in the output. May be specified 0 or more times.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--remove-filter=STRING</span></code></td>
<td>Specifies a VCF FILTER tag to remove from the output. May be specified 0 or more times.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--remove-filters</span></code></td>
<td>Set to remove all of the FILTER tags from the output.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--remove-format=STRING</span></code></td>
<td>Specifies a VCF FORMAT tag to remove from the output. May be specified 0 or more times.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--remove-info=STRING</span></code></td>
<td>Specifies a VCF INFO tag to remove from the output. May be specified 0 or more times.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--remove-infos</span></code></td>
<td>Set to remove all of the INFO tags from the output.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--remove-qual</span></code></td>
<td>Remove the QUAL field.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--remove-sample=STRING</span></code></td>
<td>Specifies a sample to remove from the output. May be specified 0 or more times.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--remove-samples</span></code></td>
<td>Set to remove all of the sample data from the output.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="11%" />
<col width="83%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-Z</span></code></td>
<td><code class="docutils literal"><span class="pre">--no-gzip</span></code></td>
<td>Set this flag to create the output file without compression. By default the output file is compressed with tabix compatible blocked gzip.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--no-index</span></code></td>
<td>Set this flag to not produce the tabix index for the output file.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Use the <code class="docutils literal"><span class="pre">vcfsubset</span></code> command to produce a smaller copy of an original VCF
file containing only the columns and information desired. For example,
to produce a VCF containing only the information for one sample from a
multiple sample VCF file use the <code class="docutils literal"><span class="pre">--keep-sample</span></code> flag to specify the
sample to keep. The various <code class="docutils literal"><span class="pre">--keep</span></code> and <code class="docutils literal"><span class="pre">--remove</span></code> options can either be
specified multiple times or with comma separated lists, for example,
<code class="docutils literal"><span class="pre">--keep-format</span> <span class="pre">GT</span> <span class="pre">--keep-format</span> <span class="pre">DP</span></code> is equivalent to <code class="docutils literal"><span class="pre">–keep-format</span> <span class="pre">GT,DP</span></code>.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#snp"><span class="std std-ref">snp</span></a>,
<a class="reference internal" href="#family"><span class="std std-ref">family</span></a>,
<a class="reference internal" href="#somatic"><span class="std std-ref">somatic</span></a>,
<a class="reference internal" href="#population"><span class="std std-ref">population</span></a>,
<a class="reference internal" href="#vcffilter"><span class="std std-ref">vcffilter</span></a>,
<a class="reference internal" href="#vcfannotate"><span class="std std-ref">vcfannotate</span></a></p>
</div>
</div>
<div class="section" id="vcfeval">
<h3>vcfeval<a class="headerlink" href="#vcfeval" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Use the <code class="docutils literal"><span class="pre">vcfeval</span></code> command to evaluate called variants for agreement with
a known baseline variant set.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfeval [OPTION]... -b FILE -c FILE -o DIR -t SDF
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfeval -b goldstandard.vcf.gz -c snps.vcf.gz -t HUMAN_reference \
  --sample daughter -f AVR -o eval
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="19%" />
<col width="75%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-b</span></code></td>
<td><code class="docutils literal"><span class="pre">--baseline=FILE</span></code></td>
<td>The VCF file containing baseline variants. For example, these may be the variants that were used to generate a synthetic
sample, a gold-standard VCF corresponding to a reference sample such as NA12878, or simply an alternative call-set being used
as a basis for comparison.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--bed-regions=FILE</span></code></td>
<td>If set, only read VCF records that overlap the ranges contained in the specified BED file.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--evaluation-regions=FILE</span></code></td>
<td>If set, evaluate within regions contained in the supplied BED file, allowing transborder matches. To be used for truth-set
high-confidence regions or other regions of interest where region boundary effects should be minimized.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-c</span></code></td>
<td><code class="docutils literal"><span class="pre">--calls=FILE</span></code></td>
<td>The VCF file containing called variants.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-o</span></code></td>
<td><code class="docutils literal"><span class="pre">--output=DIR</span></code></td>
<td>The name of the output directory.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--region=STRING</span></code></td>
<td>If set, only read VCF records that overlap the specified region. The format is one of &lt;template_name&gt;,
&lt;template_name&gt;:start-end or &lt;template_name&gt;:start+length</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-t</span></code></td>
<td><code class="docutils literal"><span class="pre">--template=SDF</span></code></td>
<td>The reference SDF on which the variants were called.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="2%" />
<col width="14%" />
<col width="84%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Filtering</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--all-records</span></code></td>
<td>Set to use all records regardless of filters. Default is to only process records where FILTER is . or PASS.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--ref-overlap</span></code></td>
<td>Allow alleles to overlap where bases of either allele are same-as-ref. (Default is to only allow VCF anchor base overlap).</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--sample=STRING</span></code></td>
<td>Set the name of the sample to select. Use the form &lt;baseline_sample&gt;,&lt;calls_sample&gt; to select different sample names for baseline and
calls. (Required when using multi-sample VCF files).</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--squash-ploidy</span></code></td>
<td>Treat heterozygous variants as homozygous ALT in both baseline and calls.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="18%" />
<col width="76%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Reporting</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--output-mode</span></code></td>
<td>Output reporting mode (Must be one of [split, annotate, combine, ga4gh, roc-only]). (Default is split).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-O</span></code></td>
<td><code class="docutils literal"><span class="pre">--sort-order=STRING</span></code></td>
<td>Set the order in which to sort the ROC scores so that &#8220;good&#8221; scores come before &#8220;bad&#8221; scores. (Must be one of [ascending,
descending]). (Default is descending).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-f</span></code></td>
<td><code class="docutils literal"><span class="pre">--vcf-score-field=STRING</span></code></td>
<td>Set the VCF format field to sort the ROC using. Also valid are &#8220;QUAL&#8221; or &#8220;INFO.&lt;name&gt;&#8221; to select the named VCF INFO field. (Default
is GQ).</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="16%" />
<col width="77%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">-Z</span></code></td>
<td><code class="docutils literal"><span class="pre">--no-gzip</span></code></td>
<td>Set this flag to create the output files without compression.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-T</span></code></td>
<td><code class="docutils literal"><span class="pre">--threads=INT</span></code></td>
<td>Specify the number of threads to use in a multi-core processor. (Default is all available cores).</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>The <code class="docutils literal"><span class="pre">vcfeval</span></code> command can be used to generate VCF files containing
called variants that were in the baseline VCF, called variants that were
not in the baseline VCF and baseline variants that were not in the
called variants. It also produces ROC curve data files based on a score
contained in a VCF field which show the predictive power of that field
for the quality of the variant calls.</p>
<p>When developing and validating sequencing pipelines and variant calling
algorithms, the comparison of variant call sets is a common problem. The
naïve way of computing these numbers is to look at the same reference
locations in the baseline (ground truth) and called variant set, and see
if genotype calls match at the same position. However, a complication
arises due to possible differences in representation for indels between
the baseline and the call sets within repeats or homopolymers, and in
multiple-nucleotide polymorphisms (MNPs), which encompass several nearby
nucleotides and are locally phased. The <code class="docutils literal"><span class="pre">vcfeval</span></code> command includes a
novel dynamic-programming algorithm for comparing variant call sets that
deals with complex call representation discrepancies, and minimizes
false positives and negatives across the entire call sets for accurate
performance evaluation. A primary advantage of <code class="docutils literal"><span class="pre">vcfeval</span></code> (compared to
other tools) is that the evaluation does not depend on normalization or
decomposition, and so the results of analysis can easily be used to
relate to the original variant calls and their annotations.</p>
<p>Note that <code class="docutils literal"><span class="pre">vcfeval</span></code> operates at the level of local haplotypes for a
sample, so for a diploid genotype, both alleles must match in order to
be considered correct. Some of the vcfeval output modes (described
below) automatically perform an additional haploid analysis phase to
identify variants which may not have a diploid match but which share a
common allele (for example, zygosity errors made during calling). If
desired, this more lenient haploid comparison can be used at the outset
by setting the <code class="docutils literal"><span class="pre">--squash-ploidy</span></code> flag (see below).</p>
<p>Note that variants selected for inclusion in a haplotype cannot be
permitted to overlap each other (otherwise the question arises of which
variant should have priority when determining the resulting haplotype),
and any well-formed call-set should not contain these situations in
order to avoid such ambiguity. When such cases are encountered by
<code class="docutils literal"><span class="pre">vcfeval</span></code>, the best non-overlapping result is determined. A special case
of overlapping variants is where calls are denoted as partially the same
as the reference (for example, a typical heterozygous call). Strictly
speaking such variants are an assertion that the relevant haplotype
bases must not be altered from the reference and overlap should not be
permitted (this is the interpretation that <code class="docutils literal"><span class="pre">vcfeval</span></code> employs by
default). However, sometimes as a result of using non-haplotype-aware
variant calling tools or when using naïve merging of multiple call sets,
a more lenient comparison is desired. The <code class="docutils literal"><span class="pre">--ref-overlap</span></code> flag will
permit such overlapping variants to both match, as long as any overlap
only occurs where one variant or other has asserted haplotype bases as
being the same as reference.</p>
<div class="section" id="haploid-matching-with-squash-ploidy">
<h4>Haploid matching with <code class="docutils literal"><span class="pre">--squash-ploidy</span></code><a class="headerlink" href="#haploid-matching-with-squash-ploidy" title="Permalink to this headline">¶</a></h4>
<p>When <code class="docutils literal"><span class="pre">--squash-ploidy</span></code> is specified, a match is attempted using <em>each</em>
of the non-reference alleles used in the sample genotype. For example if
the baseline and call VCFs each had a record with the same REF and ALT
alleles declared, the following GT fields would be considered a match:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>0/1, 1/1, 1/2
0/2, 1/2, 2/2
</pre></div>
</div>
<p>Thus a haploid match is where the baseline and calls share a common
allele.</p>
</div>
<div class="section" id="comparing-with-a-vcf-that-has-no-sample-column">
<h4>Comparing with a VCF that has no sample column<a class="headerlink" href="#comparing-with-a-vcf-that-has-no-sample-column" title="Permalink to this headline">¶</a></h4>
<p>A common scenario is to match a call set against a baseline which
contains no sample column, where the objective is to identify which
baseline alleles which have been called. One example of this is to
identify whether calls match a database of known high-priority somatic
variants such as COSMIC, or to find calls which have been previously
seen in a population allele database such as ExAC. Ordinarily
<code class="docutils literal"><span class="pre">vcfeval</span></code> requires the input VCFs to contain a sample column
containing a genotype in the GT field, however, it is possible to
specify a special sample name of &#8216;ALT&#8217; in order to indicate that the
the genotypes for comparison should be derived from the ALT alleles of
the record. This can be specified independently for baseline and calls,
for example:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfeval -t build37.sdf -b cosmic.vcf.gz -c tumor-calls.vcf.gz \
--squash-ploidy --sample ALT,tumor -o tumor-vs-cosmic
</pre></div>
</div>
<p>Which would perform a haploid matching of the GT of the called sample
&#8216;tumor&#8217; against all possible haploid genotypes in the cosmic VCF. The
resulting true positives file contains all the calls containing an
allele present in the cosmic VCF.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">It is also possible to run a diploid comparison by omitting
<code class="docutils literal"><span class="pre">--squash-ploidy</span></code>, but this is not usually required, and is
computationally more intensive since there may be many more possible
diploid genotypes to explore, particularly if the ALT VCF contains
many multiallelic records.)</p>
</div>
</div>
<div class="section" id="evaluation-with-respect-to-regions">
<h4>Evaluation with respect to regions<a class="headerlink" href="#evaluation-with-respect-to-regions" title="Permalink to this headline">¶</a></h4>
<p>When evaluating exome variant calls, it may be useful to restrict
analysis only to exome target regions.  In this case, supply a BED file
containing the list of regions to restrict analysis to via the
<code class="docutils literal"><span class="pre">--bed-regions</span></code> flag. For a quick way to restrict analysis only to a
single region, the <code class="docutils literal"><span class="pre">--region</span></code> flag is also accepted. Note that when
restricting analysis to regions, there may be variants which can not be
correctly evaluated near the borders of each analysis region, if
determination of equivalence would require inclusion of variants outside
of the region. For this reason, it is recommended that such regions be
relatively inclusive.</p>
<p>When matching against gold standard truth sets which have an
accompanying high-confidence regions BED file, the flag
<code class="docutils literal"><span class="pre">--evaluation-regions</span></code> should be used instead of <code class="docutils literal"><span class="pre">--bed-regions</span></code>, as
it has special matching semantics that aims to reduce comparison region
boundary effects. When this comparison method is used, call variants
which match a baseline variant are only considered a true positive if
the baseline variant is inside the high confidence regions, and call
variants are only considered false positive if they fall inside the high
confidence regions.</p>
</div>
<div class="section" id="vcfeval-outputs">
<h4>vcfeval outputs<a class="headerlink" href="#vcfeval-outputs" title="Permalink to this headline">¶</a></h4>
<p>The primary outputs of <code class="docutils literal"><span class="pre">vcfeval</span></code> are VCF files indicating which variants
matched between the baseline and the calls VCF, and data files
containing information used to generate ROC curves with the <code class="docutils literal"><span class="pre">rocplot</span></code>
command (or via spreadsheet). <code class="docutils literal"><span class="pre">vcfeval</span></code> supports different VCF
output modes which can be selected with the <code class="docutils literal"><span class="pre">--output-mode</span></code> flag
according to the type of analysis workflow desired. The following modes
are available:</p>
<div class="section" id="split-output-mode-split">
<h5>Split (<code class="docutils literal"><span class="pre">--output-mode=split</span></code>)<a class="headerlink" href="#split-output-mode-split" title="Permalink to this headline">¶</a></h5>
<p>This output mode is the default, and produces separate VCF files for
each of the match categories. The individual VCF records in these files
are not altered in any way, preserving all annotations present in the
input files.</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">tp.vcf</span></code> – contains those variants from the <em>calls</em> VCF which agree
with variants in the baseline VCF</li>
<li><code class="docutils literal"><span class="pre">tp-baseline.vcf</span></code> – contains those variants from the <em>baseline</em> VCF
which agree with variants in the calls VCF. Thus, the variants in
<code class="docutils literal"><span class="pre">tp.vcf</span></code> and <code class="docutils literal"><span class="pre">tp-baseline.vcf</span></code> are equivalent. This file can be used
to successively refine a highly sensitive baseline variant set to
produce a consensus from several call sets.</li>
<li><code class="docutils literal"><span class="pre">fp.vcf</span></code> – contains variants from the <em>calls</em> VCF which do not agree
with baseline variants.</li>
<li><code class="docutils literal"><span class="pre">fn.vcf</span></code> – contains variants from the <em>baseline</em> VCF which were not
correctly called.</li>
</ul>
<p>This mode performs a single pass comparison, either in diploid mode (the
default), or haploid mode (if <code class="docutils literal"><span class="pre">--squash-ploidy</span></code> has been set). The
separate output files produced by this mode allow the use of <code class="docutils literal"><span class="pre">vcfeval</span></code>
as an advanced haplotype-aware VCF intersection tool.</p>
</div>
<div class="section" id="annotate-output-mode-annotate">
<h5>Annotate (<code class="docutils literal"><span class="pre">--output-mode=annotate</span></code>)<a class="headerlink" href="#annotate-output-mode-annotate" title="Permalink to this headline">¶</a></h5>
<p>This output mode does not split the input VCFs by match status, but
instead adds <code class="docutils literal"><span class="pre">INFO</span></code> annotations containing the match status of each
record:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">calls.vcf</span></code> – contains variants from the <em>calls</em> VCF, augmented with
match status annotations.</li>
<li><code class="docutils literal"><span class="pre">baseline.vcf</span></code> – contains variants from the <em>baseline</em> VCF, augmented
with match status annotations.</li>
</ul>
<p>This output mode automatically performs two comparison passes, the first
finds diploid matches (assigned a match status of <code class="docutils literal"><span class="pre">TP</span></code>), and a second
pass that applies a haploid mode to the false positives and false
negatives in order to find calls (such as zygosity errors) that contain
a common allele. This second category of match are annotated with status
<code class="docutils literal"><span class="pre">FN_CA</span></code> or <code class="docutils literal"><span class="pre">FP_CA</span></code> in the output VCFs, and those calls which do not
have any match are assigned status <code class="docutils literal"><span class="pre">FN</span></code> or <code class="docutils literal"><span class="pre">FP</span></code>. A status value of
<code class="docutils literal"><span class="pre">IGN</span></code> indicates a VCF record which was ignored (for example, due to
having a non-PASS filter status, representing a structural variant, or
otherwise containing a non-variant genotype). A status of <code class="docutils literal"><span class="pre">OUT</span></code>
indicates a VCF record which does not contain a match status due to
falling outside the evaluation regions when <code class="docutils literal"><span class="pre">--evaluation-regions</span></code> is
being used.</p>
</div>
<div class="section" id="combine-output-mode-combine">
<h5>Combine (&#8211;output-mode=combine)<a class="headerlink" href="#combine-output-mode-combine" title="Permalink to this headline">¶</a></h5>
<p>This output mode provides an easy way to view the baseline and call
variants in a single two-sample VCF.</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">output.vcf</span></code> – contains variants from both the <em>baseline</em> and <em>calls</em>
VCFs, augmented with match status annotations. The sample under
comparison from each of the input VCFs is extracted as a column in
the output. As the VCF records from the baseline and calls typically
have very different input annotations which can be difficult to
merge, and to keep the output format simple, there is no attempt to
preserve any of the original variant annotations.</li>
</ul>
<p>As with the annotation output mode, this output mode automatically
performs two comparison passes to find both diploid matches and haploid
(lenient) matches.</p>
</div>
<div class="section" id="roc-only-output-mode-roc-only">
<h5>ROC-only (&#8211;output-mode=roc-only)<a class="headerlink" href="#roc-only-output-mode-roc-only" title="Permalink to this headline">¶</a></h5>
<p>This output mode provides a lightweight way to run performance
benchmarking, as VCF file output is omitted, and only ROC data files are
produced.</p>
<p>All of the output modes produce the following ROC data files:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">weighted_roc.tsv</span></code> – contains ROC data derived from all analyzed
call variants, regardless of their representation. Columns include
the score field, and standard accuracy metrics such as true
positives, false positives, false negatives, precision, sensitivity,
and f-measure corresponding to each score threshold.</li>
<li><code class="docutils literal"><span class="pre">snp_roc.tsv</span></code> – contains ROC data derived from only those call
variants which were represented as SNPs. This file includes a subset
of accuracy metrics, as the computation of some metrics is not
meaningful on a subset of the data where representation may differ
between the baseline and the call.</li>
<li><code class="docutils literal"><span class="pre">non_snp_roc.tsv</span></code> – contains ROC data derived from only those call
variants which were not represented as SNPs. As above, not all
metrics are computed for this file.</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">In addition, <code class="docutils literal"><span class="pre">vcfeval</span></code> has an output mode
(<code class="docutils literal"><span class="pre">--output-mode=ga4gh</span></code>) which produces the intermediate evaluation
format defined by the GA4GH Benchmarking Team, without additional
statistics files. This mode is not generally intended for end users,
rather it is used when vcfeval is selected as the comparison engine
inside the hap.py benchmarking tool see:
<a class="reference external" href="https://github.com/ga4gh/benchmarking-tools">https://github.com/ga4gh/benchmarking-tools</a> and
<a class="reference external" href="https://github.com/Illumina/hap.py">https://github.com/Illumina/hap.py</a></p>
</div>
<p>Multiple ROC data files (from a single or several <code class="docutils literal"><span class="pre">vcfeval</span></code> runs) can
be plotted with the <code class="docutils literal"><span class="pre">rocplot</span></code> command, which allows output to a PNG or
SVG image or analysis in an interactive GUI that provides zooming and
visualization of the effects of threshold adjustment. As these files are
simple Tab-Separated-Value format, they can also be loaded into a
spreadsheet tool or processed with shell scripts.</p>
<p>While ROC curve analysis provides a much more thorough method for
examining the performance of a call set with respect to a baseline truth
set, for convenience, <code class="docutils literal"><span class="pre">vcfeval</span></code> also produces a <code class="docutils literal"><span class="pre">summary.txt</span></code> file
which indicates match summary statistics that correspond to two key
points on the ROC curve.  The first point is where all called variants
are included (i.e. no thresholding on a score value); and second point
corresponding to a score threshold that maximises the F-measure of the
curve. While this latter point is somewhat arbitrary, it represents a
balanced tradeoff between precision and sensitivity which is likely to
provide a fairer comparison when comparing call sets from different
callers.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#snp"><span class="std std-ref">snp</span></a>,
<a class="reference internal" href="#popsim"><span class="std std-ref">popsim</span></a>,
<a class="reference internal" href="#samplesim"><span class="std std-ref">samplesim</span></a>,
<a class="reference internal" href="#childsim"><span class="std std-ref">childsim</span></a>,
<a class="reference internal" href="#rocplot"><span class="std std-ref">rocplot</span></a></p>
</div>
</div>
</div>
</div>
<div class="section" id="pedfilter">
<h3>pedfilter<a class="headerlink" href="#pedfilter" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Filter and convert a pedigree file.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg pedfilter [OPTION]... FILE
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg pedfilter --remove-parentage mypedigree.ped
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="4%" />
<col width="13%" />
<col width="83%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td>FILE</td>
<td>The pedigree file to process, may be PED or VCF, use &#8216;-&#8216; to read from stdin.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="3%" />
<col width="22%" />
<col width="75%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Filtering</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--keep-primary</span></code></td>
<td>Keep only primary individuals (those with a PED individual line / VCF sample column).</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--remove-parentage</span></code></td>
<td>Remove all parent-child relationship information.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="15%" />
<col width="80%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Reporting</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--vcf</span></code></td>
<td>Output pedigree in in the form of a VCF header rather than PED.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="21%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>The <code class="docutils literal"><span class="pre">pedfilter</span></code> comand can be used to perform manipulations on pedigree
information and convert pedigree information between PED and VCF header
format.</p>
<p>The VCF files output by the <code class="docutils literal"><span class="pre">family</span></code> and <code class="docutils literal"><span class="pre">population</span></code> commands contain
full pedigree information represented as VCF header lines, and the
<code class="docutils literal"><span class="pre">pedfilter</span></code> command allows this information to be extracted in PED
format.</p>
<p>This command produces the pedigree output on standard output, which can
be redirected to a file or another pipeline command as required.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#family"><span class="std std-ref">family</span></a>,
<a class="reference internal" href="#population"><span class="std std-ref">population</span></a>,
<a class="reference internal" href="#mendelian"><span class="std std-ref">mendelian</span></a>,
<a class="reference internal" href="#pedstats"><span class="std std-ref">pedstats</span></a></p>
</div>
</div>
<div class="section" id="pedstats">
<h3>pedstats<a class="headerlink" href="#pedstats" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Output information from pedigree files of various formats.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg pedstats [OPTION]... FILE
</pre></div>
</div>
<p><strong>Example:</strong></p>
<p>For a summary of pedigree information:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg pedstats ceph_pedigree.ped

Pedigree file: /data/ceph/ceph_pedigree.ped

Total samples:               17
Primary samples:             17
Male samples:                 9
Female samples:               8
Afflicted samples:            0
Founder samples:              4
Parent-child relationships:  26
Other relationships:          0
Families:                     3
</pre></div>
</div>
<p>For quick pedigree visualization using graphviz and ImageMagick, use a
command-line such as:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ dot -Tpng &lt;(rtg pedstats --dot &quot;A Title&quot; mypedigree.ped) | display -
</pre></div>
</div>
<p>For a larger pedigree:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ dot -Tpdf -o mypedigree.pdf &lt;(rtg pedstats --dot &quot;Study&quot; mypedigree.ped)
</pre></div>
</div>
<p>To output a list of all founders:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg pedstats --founder-ids ceph_pedigree.ped
NA12889
NA12890
NA12891
NA12892
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="4%" />
<col width="13%" />
<col width="83%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td>FILE</td>
<td>The pedigree file to process, may be PED or VCF, use &#8216;-&#8216; to read from stdin.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="4%" />
<col width="22%" />
<col width="74%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Reporting</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--dot=STRING</span></code></td>
<td>Output pedigree in GraphViz format, using the supplied text as a title.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--families</span></code></td>
<td>Output information about family structures.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--female-ids</span></code></td>
<td>Output ids of all females.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--founder-ids</span></code></td>
<td>Output ids of all founders.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--male-ids</span></code></td>
<td>Output ids of all males.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--maternal-ids</span></code></td>
<td>Output ids of maternal individuals.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--paternal-ids</span></code></td>
<td>Output ids of paternal individuals.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--primary-ids</span></code></td>
<td>Output ids of all primary individuals.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="21%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Used to show pedigree summary statistics or select groups of individual
Ids. In particular, it is possible to generate a simple pedigree
visualization.</p>
<p>The VCF files output by the <code class="docutils literal"><span class="pre">family</span></code> and <code class="docutils literal"><span class="pre">population</span></code> commands contain
full pedigree information represented as VCF header lines, and the
<code class="docutils literal"><span class="pre">pedstats</span></code> command can also take these VCFs as input.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#family"><span class="std std-ref">family</span></a>,
<a class="reference internal" href="#population"><span class="std std-ref">population</span></a>,
<a class="reference internal" href="#pedfilter"><span class="std std-ref">pedfilter</span></a></p>
</div>
</div>
<div class="section" id="rocplot">
<h3>rocplot<a class="headerlink" href="#rocplot" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>Plot ROC curves from <code class="docutils literal"><span class="pre">readsimeval</span></code> and <code class="docutils literal"><span class="pre">vcfeval</span></code> ROC data files, either
to an image, or using an interactive GUI.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg rocplot [OPTION]... FILE+
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg rocplot [OPTION]... --curve STRING
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg rocplot eval/weighted_roc.tsv.gz
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="3%" />
<col width="16%" />
<col width="81%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">File Input/Output</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--curve=STRING</span></code></td>
<td>Specify a ROC data file with title optionally specified (path[=title]). May be specified 0 or more times.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--png=FILE</span></code></td>
<td>Set to output a PNG image to the given file instead of loading the interactive plot.</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--svg=FILE</span></code></td>
<td>Set to output a SVG image to the given file instead of loading the interactive plot.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">FILE+</span></code></td>
<td>Specify the ROC data file to plot. May be specified 0 or more times.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="11%" />
<col width="32%" />
<col width="58%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Reporting</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--hide-sidepane</span></code></td>
<td>Set to hide the sidepane from the GUI on startup.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--line-width=INT</span></code></td>
<td>Set the line width for the plots. (Default is 2).</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--scores</span></code></td>
<td>Set to show scores on the plot.</td>
</tr>
<tr class="row-odd"><td>&nbsp;</td>
<td><code class="docutils literal"><span class="pre">--precision-sensitivity</span></code></td>
<td>Produce a precision/sensitivity plot instead of ROC</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-t</span></code></td>
<td><code class="docutils literal"><span class="pre">--title=STRING</span></code></td>
<td>Set the title for the plot.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="21%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">Utility</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">-h</span></code></td>
<td><code class="docutils literal"><span class="pre">--help</span></code></td>
<td>Prints help on command-line flag usage.</td>
</tr>
</tbody>
</table>
<p><strong>Usage:</strong></p>
<p>Used to produce ROC plots from the ROC files produced by <code class="docutils literal"><span class="pre">readsimeval</span></code>
and <code class="docutils literal"><span class="pre">vcfeval</span></code>. By default this opens the ROC plots in an interactive
viewer. On a system with only console access the plot can be saved
directly to an image file using the either the <code class="docutils literal"><span class="pre">--png</span></code> or <code class="docutils literal"><span class="pre">--svg</span></code> parameter.</p>
<p>Some quick tips for the interactive GUI:</p>
<ul class="simple">
<li>Select regions within the graph to zoom in. Right click to bring up a
context menu that allows resetting the zoom.</li>
<li>Click on a spot in the graph to show the equivalent accuracy metrics
for that location in the status bar. Clicking to the left or below
the axes will clear the cross-hair. Note that sensitivity depends on
the baseline total number of variants being correct. If for example
the ROC curve corresponds to evaluating an exome call-set against a
whole-genome baseline, this number will be inaccurate.</li>
<li>Additional ROC data files can be loaded by clicking on the &#8216;Open...&#8217;
button.</li>
<li>Each ROC curve can be shown/hidden, renamed, and reordered in it&#8217;s
widget area on the right hand side of the UI.</li>
<li>Each ROC curve has a slider to simulate the effect of applying a
threshold on the scoring attribute. If the &#8216;show scores&#8217; option is
set, this provides an easy way to select appropriate filter threshold
values.</li>
<li>The &#8216;Cmd&#8217; button will print to the console a command-line which is
equivalent to the currently displayed set of ROC curves, which gives
an easy way to replicate the current set of curves in another
session.</li>
<li>There is a drop down that allows for switching between ROC and
precision/sensitivity graph types.</li>
</ul>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#readsimeval"><span class="std std-ref">readsimeval</span></a>,
<a class="reference internal" href="#vcfeval"><span class="std std-ref">vcfeval</span></a></p>
</div>
</div>
<div class="section" id="version">
<h3>version<a class="headerlink" href="#version" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>The RTG version display utility.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg version
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg version

Product: RTG Core 3.5
Core Version: 4586490 2015-12-04
RAM: 3.5GB of 3.8GB RAM can be used by RTG (91%)
CPU: Defaulting to 4 of 4 available processors (100%)
License: Expires on 2016-03-30
Contact: support@realtimegenomics.com

Patents / Patents pending:
US: 7,640,256, 13/129,329, 13/681,046, 13/681,215, 13/848,653,
13/925,704, 14/015,295, 13/971,654, 13/971,630, 14/564,810
UK: 1222923.3, 1222921.7, 1304502.6, 1311209.9, 1314888.7, 1314908.3
New Zealand: 626777, 626783, 615491, 614897, 614560
Australia: 2005255348, Singapore: 128254

Citation:

John G. Cleary, Ross Braithwaite, Kurt Gaastra, Brian S. Hilbush, Stuart Inglis, Sean A. Irvine, Alan Jackson, Richard Littin, Sahar Nohzadeh-Malakshah, Mehul Rathod, David Ware, Len Trigg, and Francisco M. De La Vega. &quot;Joint Variant and De Novo Mutation Identification on Pedigrees from High-Throughput Sequencing Data.&quot; Journal of Computational Biology. June 2014, 21(6): 405-419.  doi:10.1089/cmb.2014.0029.

(c) Real Time Genomics, 2014
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<p>There are no options associated with the <code class="docutils literal"><span class="pre">version</span></code> command.</p>
<p><strong>Usage:</strong></p>
<p>Use the <code class="docutils literal"><span class="pre">version</span></code> command to display release and version information.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#help"><span class="std std-ref">help</span></a>,
<a class="reference internal" href="#license"><span class="std std-ref">license</span></a></p>
</div>
</div>
<div class="section" id="license">
<h3>license<a class="headerlink" href="#license" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>The RTG license display utility.</p>
<p><strong>Syntax:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg license
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg license
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<p>There are no options associated with the license command.</p>
<p><strong>Usage:</strong></p>
<p>Use the <code class="docutils literal"><span class="pre">license</span></code> command to display license information and expiration
date. Output at the command line (standard output) shows command name,
licensed status, and command release level.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#help"><span class="std std-ref">help</span></a>,
<a class="reference internal" href="#version"><span class="std std-ref">version</span></a></p>
</div>
</div>
<div class="section" id="help">
<h3>help<a class="headerlink" href="#help" title="Permalink to this headline">¶</a></h3>
<p><strong>Synopsis:</strong></p>
<p>The RTG help command provides online help for all RTG commands.</p>
<p><strong>Syntax:</strong></p>
<p>List all commands:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg help
</pre></div>
</div>
<p>Show usage syntax and flags for one command:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg help COMMAND
</pre></div>
</div>
<p><strong>Example:</strong></p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg help format
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<p>There are no options associated with the <code class="docutils literal"><span class="pre">help</span></code> command.</p>
<p><strong>Usage:</strong></p>
<p>Use the <code class="docutils literal"><span class="pre">help</span></code> command to view syntax and usage information for the main
<code class="docutils literal"><span class="pre">rtg</span></code> command as well as individual RTG commands.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#license"><span class="std std-ref">license</span></a>,
<a class="reference internal" href="#version"><span class="std std-ref">version</span></a></p>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="product_usage_wrapper.html" title="RTG product usage - baseline progressions"
             >next</a> |</li>
        <li class="right" >
          <a href="overview.html" title="Overview"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">RTG Tools Operations Manual v3.7</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2016, Real Time Genomics.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.6.
    </div>
  </body>
</html>